<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • updog</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Changelog" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/multidog.html">Genotyping Many SNP's with multidog()</a>
    </li>
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/dcgerard/updog/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href='http://github.com/dcgerard/updog/blob/master/NEWS.md'><code>NEWS.md</code></a></small>
    </div>

    <div id="updog-200" class="section level1">
<h1 class="page-header" data-toc-text="2.0.0">
<a href="#updog-200" class="anchor"></a>updog 2.0.0<small> Unreleased </small>
</h1>
<p>This is a massive edit of the updog software. Major changes include:</p>
<ol>
<li>No more support of <code>model = "ash"</code>. It seemed that <code>model = "norm"</code> was always better and faster, so I just got rid of the <code>"ash"</code> option. This also extremely simplified the code.</li>
<li>Removal of <code>mupdog()</code>. I think this was a good idea, but the computation was way too slow to be usable.</li>
<li>Revision of <code>model = "f1pp"</code> and <code>model = "s1pp"</code>. These now include interpretable parameterizations that are meant to be identified via another R package. But support is only for tetraploids right now.</li>
<li>
<code><a href="../reference/multidog.html">multidog()</a></code> now prints some nice ASCII art when it’s run.</li>
</ol>
</div>
    <div id="updog-121" class="section level1">
<h1 class="page-header" data-toc-text="1.2.1">
<a href="#updog-121" class="anchor"></a>updog 1.2.1<small> Unreleased </small>
</h1>
<ul>
<li>Adds <code><a href="../reference/filter_snp.html">filter_snp()</a></code> for filtering the output of <code><a href="../reference/multidog.html">multidog()</a></code> based on predicates in terms of the variables in <code>snpdf</code>.</li>
<li>Removes stringr from Imports. I was only using it in one place so I replaced that code with base R code.</li>
<li>Removes Rmpfr from Suggests. No longer needed since CVXR is no longer suggested.</li>
</ul>
</div>
    <div id="updog-120" class="section level1">
<h1 class="page-header" data-toc-text="1.2.0">
<a href="#updog-120" class="anchor"></a>updog 1.2.0<small> 2020-01-28 </small>
</h1>
<ul>
<li>Adds <code><a href="../reference/multidog.html">multidog()</a></code> for genotypying multiple SNP’s using parallel computing.</li>
<li>Adds <code><a href="../reference/plot.multidog.html">plot.multidog()</a></code> for plotting the output of <code><a href="../reference/multidog.html">multidog()</a></code>.</li>
<li>Adds <code><a href="../reference/format_multidog.html">format_multidog()</a></code> for formatting the output of <code><a href="../reference/multidog.html">multidog()</a></code> to be a matrix.</li>
<li>Removes dependency on CVXR. This makes install and maintenance a little easier. The defaults for this specific problem were a little faster anyway.</li>
<li>No longer changes the color scale in <code><a href="../reference/plot_geno.html">plot_geno()</a></code> based on what genotypes are present.</li>
<li>In .cpp files, we now coerce objects to be unsigned before comparing. This gets rid of some warnings during install.</li>
</ul>
</div>
    <div id="updog-113" class="section level1">
<h1 class="page-header" data-toc-text="1.1.3">
<a href="#updog-113" class="anchor"></a>updog 1.1.3<small> 2019-11-21 </small>
</h1>
<ul>
<li>Updates documentation to include the Bioinformatics publication, Gerard and Ferrão (2019) &lt;<a href="https://doi.org/10.1093/bioinformatics/btz852">doi:10.1093/bioinformatics/btz852</a>&gt;.</li>
<li>Adds the “internal” keyword to functions that most users don’t need.</li>
<li>Removes the tidyverse from the Suggests field. I was only using this in the vignettes, so I changed it to base R (except for ggplot2).</li>
</ul>
</div>
    <div id="updog-111" class="section level1">
<h1 class="page-header" data-toc-text="1.1.1">
<a href="#updog-111" class="anchor"></a>updog 1.1.1<small> 2019-09-09 </small>
</h1>
<ul>
<li>Updates documentation to include Gerard and Ferrão (2019) &lt;<a href="https://doi.org/10.1101/751784">doi:10.1101/751784</a>&gt; as a reference.</li>
<li>Minor fixes to documentation.</li>
</ul>
</div>
    <div id="updog-110" class="section level1">
<h1 class="page-header" data-toc-text="1.1.0">
<a href="#updog-110" class="anchor"></a>updog 1.1.0<small> 2019-07-31 </small>
</h1>
<ul>
<li>Introduces more flexible priors for more general populations.</li>
<li>Places a normal prior distribution on the logit of the overdispersion parameter. This might change genotype calls from previous versions of updog. To reproduce the genotype calls from previous versions of updog, simply set <code>mean_od = 0</code> and <code>var_od = Inf</code> in <code><a href="../reference/flexdog.html">flexdog()</a></code>.</li>
<li>Adds the <code>method = "custom"</code> option to <code><a href="../reference/flexdog.html">flexdog()</a></code>. This lets users choose the genotype distribution if it is completely known a priori.</li>
<li>Documentation updates.</li>
</ul>
</div>
    <div id="updog-101" class="section level1">
<h1 class="page-header" data-toc-text="1.0.1">
<a href="#updog-101" class="anchor"></a>updog 1.0.1<small> 2018-07-27 </small>
</h1>
<ul>
<li><p>Fixes a bug with option <code>model = "s1pp"</code> in <code><a href="../reference/flexdog.html">flexdog()</a></code>. I was originally not constraining the levels of preferential pairing to be the same in both segregations of the same parent. This is now fixed. But the downside is that <code>model = "s1pp"</code> is now only supported for <code>ploidy = 4</code> or <code>ploidy = 6</code>. This is because the optimization becomes more difficult for larger ploidy levels.</p></li>
<li>
<p>I fixed some documentation. Perhaps the biggest error comes from this snippet from the original documentation of <code>flexdog</code>:</p>
<blockquote>
<p>The value of <code>prop_mis</code> is a very intuitive measure for the quality of the SNP. <code>prop_mis</code> is the posterior proportion of individuals mis-genotyped. So if you want only SNPS that accurately genotype, say, 95% of the individuals, you could discard all SNPs with a <code>prop_mis</code> under 0.95.</p>
</blockquote>
<p>This now says</p>
<blockquote>
<p>The value of prop_mis is a very intuitive measure for the quality of the SNP. prop_mis is the posterior proportion of individuals mis-genotyped. So if you want only SNPS that accurately genotype, say, 95% of the individuals, you could discard all SNPs with a prop_mis <strong>over 0.05</strong>.</p>
</blockquote>
</li>
<li><p>I’ve now exported some C++ functions that I think are useful. You can call them in the usual way: <a href="http://r-pkgs.had.co.nz/src.html#cpp-import" class="uri">http://r-pkgs.had.co.nz/src.html#cpp-import</a>.</p></li>
</ul>
</div>
    <div id="updog-0990" class="section level1">
<h1 class="page-header" data-toc-text="0.99.0">
<a href="#updog-0990" class="anchor"></a>updog 0.99.0<small> Unreleased </small>
</h1>
<ul>
<li>This is a complete re-working of the code in <code>updog</code>. The old version may be found in the <a href="https://github.com/dcgerard/updogAlpha"><code>updogAlpha</code> package</a>.</li>
<li>The main function is now <code><a href="../reference/flexdog.html">flexdog()</a></code>.</li>
<li>An experimental approach <code>mupdog()</code> is now live. We provide no guarantees about <code>mupdog()</code>’s performance.</li>
<li>Oracle misclassification error rates may be calculated in <code><a href="../reference/oracle_mis.html">oracle_mis()</a></code>.</li>
<li>Genotypes can be simulated using <code><a href="../reference/rgeno.html">rgeno()</a></code>.</li>
<li>Next-generation sequencing data can be simulated using <code><a href="../reference/rflexdog.html">rflexdog()</a></code>.</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


