<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • updog</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><meta property="og:image" content="/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">updog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a></li>
    <li><a class="dropdown-item" href="../articles/oracle_calculations.html">Oracle Calculations</a></li>
    <li><a class="dropdown-item" href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a></li>
    <li><a class="dropdown-item" href="../articles/smells_like_updog.html">Example Use of Updog</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/dcgerard/updog/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.2.0" id="updog-220">Updog 2.2.0<a class="anchor" aria-label="anchor" href="#updog-220"></a></h2>
<ul><li>Added a prior for random mating in <code><a href="../reference/flexdog.html">flexdog()</a></code> and <code><a href="../reference/multidog.html">multidog()</a></code>. Use <code>model = "rm"</code> to implement.</li>
<li>Added a method to use genotype likelihoods to estimate the site-frequency spectrum using the algorithm of Li (2011) <a href="doi:10.1093/bioinformatics/btr509" class="uri">doi:10.1093/bioinformatics/btr509</a>
</li>
<li>Updated the penalty on the bias parameter to result in more symmetric genotyping in edge cases.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.1.5" id="updog-215">updog 2.1.5<a class="anchor" aria-label="anchor" href="#updog-215"></a></h2><p class="text-muted">CRAN release: 2023-11-29</p>
<ul><li>Used updated Rcpp to generate RcppExports to fix a CRAN warning.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.1.4" id="updog-214">updog 2.1.4<a class="anchor" aria-label="anchor" href="#updog-214"></a></h2><p class="text-muted">CRAN release: 2023-11-17</p>
<ul><li>Removed <a href="https://jrnold.github.io/ggthemes/" class="external-link">ggthemes</a> dependency.</li>
<li>Removed usage of <code><a href="https://ggplot2.tidyverse.org/reference/aes_.html" class="external-link">ggplot2::aes_string()</a></code>, since it is deprecated, and replaced it with tidy evaluation idioms.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.1.3" id="updog-213">updog 2.1.3<a class="anchor" aria-label="anchor" href="#updog-213"></a></h2><p class="text-muted">CRAN release: 2022-10-18</p>
<ul><li>Bug fix: Use <code>&amp;&amp;</code> instead of <code>&amp;</code> in C++.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.1.2" id="updog-212">updog 2.1.2<a class="anchor" aria-label="anchor" href="#updog-212"></a></h2><p class="text-muted">CRAN release: 2022-01-24</p>
<ul><li>
<p>Fixed a bug in my use of <code><a href="https://rdrr.io/pkg/assertthat/man/are_equal.html" class="external-link">assertthat::are_equal()</a></code> and <code><a href="https://testthat.r-lib.org/reference/equality-expectations.html" class="external-link">testthat::expect_equal()</a></code>. See 21 Jan 2022 <a href="https://developer.r-project.org/blosxom.cgi/R-devel/NEWS" class="external-link">R-devel/NEWS</a> where it states:</p>
<blockquote>
<p><code><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal.numeric()</a></code> gains a sanity check on its <code>tolerance</code> argument - calling <code>all.equal(a, b, c)</code> for three numeric vectors is a surprisingly common error.</p>
</blockquote>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.1.1" id="updog-211">updog 2.1.1<a class="anchor" aria-label="anchor" href="#updog-211"></a></h2><p class="text-muted">CRAN release: 2021-10-25</p>
<ol style="list-style-type: decimal"><li>Added an upper bound to the sequencing error rate in <code><a href="../reference/flexdog_full.html">flexdog_full()</a></code> (and, hence, <code><a href="../reference/flexdog.html">flexdog()</a></code> and <code><a href="../reference/multidog.html">multidog()</a></code>). This protects against some poor behavior observed in a corner case. Specifically, F1 populations where the offspring are all the same genotype and is sequenced at moderate to low depth.</li>
<li>Fixed some stale URLs, fixed some style issues found by lintr.</li>
</ol></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.1.0" id="updog-210">updog 2.1.0<a class="anchor" aria-label="anchor" href="#updog-210"></a></h2>
<ol style="list-style-type: decimal"><li>The parallel backend in <code><a href="../reference/multidog.html">multidog()</a></code> is now handled by the <a href="https://cran.r-project.org/package=future" class="external-link"><code>future</code></a> package. If you use the <code>nc</code> argument in <code><a href="../reference/multidog.html">multidog()</a></code>, it should still run in parallel using multiple R sessions on your local machine. However, you can now use the functionality of <code>future</code> to choose your own evaluation strategy, after setting <code>nc = NA</code>. This will also allow you to use schedulers in high performance computing environments through the <a href="https://cran.r-project.org/package=future.batchtools" class="external-link"><code>future.batchtools</code></a> package. See the <code><a href="../reference/multidog.html">multidog()</a></code> function documentation for more details.</li>
<li>The vignette “Genotyping Many SNPs with multidog()” goes through an example of using the <code>future</code> package.</li>
<li>A new experimental function, <code>export_vcf()</code>, is in the works to export <code>multidog</code> objects to a VCF file. This is not yet exported because I still have a few bugs to fix.</li>
<li>
<code><a href="../reference/plot.multidog.html">plot.multidog()</a></code> will now plot the parent read-counts in F1 and S1 populations.</li>
<li>Internally, <code><a href="../reference/multidog.html">multidog()</a></code> now uses iterators through the <a href="https://cran.r-project.org/package=iterators" class="external-link"><code>iterators</code></a> package to send only subsets of the data to each R process.</li>
<li>A new internal <code>.combine</code> function is used in the <code>foreach()</code> call of <code><a href="../reference/multidog.html">multidog()</a></code> in order to decrease the memory usage of <code><a href="../reference/multidog.html">multidog()</a></code>.</li>
</ol></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.0.2" id="updog-202">updog 2.0.2<a class="anchor" aria-label="anchor" href="#updog-202"></a></h2><p class="text-muted">CRAN release: 2020-07-21</p>
<p>This is a massive edit of the updog software. Major changes include:</p>
<ol style="list-style-type: decimal"><li>No more support of <code>model = "ash"</code>. It seemed that <code>model = "norm"</code> was always better and faster, so I just got rid of the <code>"ash"</code> option. This also extremely simplified the code.</li>
<li>Removal of <code>mupdog()</code>. I think this was a good idea, but the computation was way too slow to be usable.</li>
<li>Revision of <code>model = "f1pp"</code> and <code>model = "s1pp"</code>. These now include interpretable parameterizations that are meant to be identified via another R package. But support is only for tetraploids right now.</li>
<li>
<code><a href="../reference/multidog.html">multidog()</a></code> now prints some nice ASCII art when it’s run.</li>
<li>
<code><a href="../reference/format_multidog.html">format_multidog()</a></code> now allows you to format multiple variables in terms of a multidimensional array.</li>
<li>Fixes a bug where <code><a href="../reference/format_multidog.html">format_multidog()</a></code> was reordering the SNP dimensions. This was fine as long as folks used dimnames properly, but now it should allow folks to also use dim positions.</li>
<li>Updog now returns genotype log-likelihoods.</li>
</ol></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.2.1" id="updog-121">updog 1.2.1<a class="anchor" aria-label="anchor" href="#updog-121"></a></h2>
<ul><li>Adds <code><a href="../reference/filter_snp.html">filter_snp()</a></code> for filtering the output of <code><a href="../reference/multidog.html">multidog()</a></code> based on predicates in terms of the variables in <code>snpdf</code>.</li>
<li>Removes stringr from Imports. I was only using it in one place so I replaced that code with base R code.</li>
<li>Removes Rmpfr from Suggests. No longer needed since CVXR is no longer suggested.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.2.0" id="updog-120">updog 1.2.0<a class="anchor" aria-label="anchor" href="#updog-120"></a></h2><p class="text-muted">CRAN release: 2020-01-28</p>
<ul><li>Adds <code><a href="../reference/multidog.html">multidog()</a></code> for genotyping multiple SNPs using parallel computing.</li>
<li>Adds <code><a href="../reference/plot.multidog.html">plot.multidog()</a></code> for plotting the output of <code><a href="../reference/multidog.html">multidog()</a></code>.</li>
<li>Adds <code><a href="../reference/format_multidog.html">format_multidog()</a></code> for formatting the output of <code><a href="../reference/multidog.html">multidog()</a></code> to be a matrix.</li>
<li>Removes dependency on CVXR. This makes install and maintenance a little easier. The defaults for this specific problem were a little faster anyway.</li>
<li>No longer changes the color scale in <code><a href="../reference/plot_geno.html">plot_geno()</a></code> based on what genotypes are present.</li>
<li>In .cpp files, we now coerce objects to be unsigned before comparing. This gets rid of some warnings during install.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.3" id="updog-113">updog 1.1.3<a class="anchor" aria-label="anchor" href="#updog-113"></a></h2><p class="text-muted">CRAN release: 2019-11-21</p>
<ul><li>Updates documentation to include the Bioinformatics publication, Gerard and Ferrão (2020) &lt;<a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link">doi:10.1093/bioinformatics/btz852</a>&gt;.</li>
<li>Adds the “internal” keyword to functions that most users don’t need.</li>
<li>Removes the tidyverse from the Suggests field. I was only using this in the vignettes, so I changed it to base R (except for ggplot2).</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.1" id="updog-111">updog 1.1.1<a class="anchor" aria-label="anchor" href="#updog-111"></a></h2><p class="text-muted">CRAN release: 2019-09-09</p>
<ul><li>Updates documentation to include Gerard and Ferrão (2020) &lt;<a href="https://doi.org/10.1101/751784" class="external-link">doi:10.1101/751784</a>&gt; as a reference.</li>
<li>Minor fixes to documentation.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.0" id="updog-110">updog 1.1.0<a class="anchor" aria-label="anchor" href="#updog-110"></a></h2><p class="text-muted">CRAN release: 2019-07-31</p>
<ul><li>Introduces more flexible priors for more general populations.</li>
<li>Places a normal prior distribution on the logit of the overdispersion parameter. This might change genotype calls from previous versions of updog. To reproduce the genotype calls from previous versions of updog, simply set <code>mean_od = 0</code> and <code>var_od =   Inf</code> in <code><a href="../reference/flexdog.html">flexdog()</a></code>.</li>
<li>Adds the <code>method = "custom"</code> option to <code><a href="../reference/flexdog.html">flexdog()</a></code>. This lets users choose the genotype distribution if it is completely known a priori.</li>
<li>Documentation updates.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.1" id="updog-101">updog 1.0.1<a class="anchor" aria-label="anchor" href="#updog-101"></a></h2><p class="text-muted">CRAN release: 2018-07-27</p>
<ul><li><p>Fixes a bug with option <code>model = "s1pp"</code> in <code><a href="../reference/flexdog.html">flexdog()</a></code>. I was originally not constraining the levels of preferential pairing to be the same in both segregations of the same parent. This is now fixed. But the downside is that <code>model = "s1pp"</code> is now only supported for <code>ploidy = 4</code> or <code>ploidy = 6</code>. This is because the optimization becomes more difficult for larger ploidy levels.</p></li>
<li>
<p>I fixed some documentation. Perhaps the biggest error comes from this snippet from the original documentation of <code>flexdog</code>:</p>
<blockquote>
<p>The value of <code>prop_mis</code> is a very intuitive measure for the quality of the SNP. <code>prop_mis</code> is the posterior proportion of individuals mis-genotyped. So if you want only SNPS that accurately genotype, say, 95% of the individuals, you could discard all SNPs with a <code>prop_mis</code> under 0.95.</p>
</blockquote>
<p>This now says</p>
<blockquote>
<p>The value of prop_mis is a very intuitive measure for the quality of the SNP. prop_mis is the posterior proportion of individuals mis-genotyped. So if you want only SNPS that accurately genotype, say, 95% of the individuals, you could discard all SNPs with a prop_mis <strong>over 0.05</strong>.</p>
</blockquote>
</li>
<li><p>I’ve now exported some C++ functions that I think are useful. You can call them in the usual way.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.0" id="updog-0990">updog 0.99.0<a class="anchor" aria-label="anchor" href="#updog-0990"></a></h2>
<ul><li>This is a complete re-working of the code in <code>updog</code>. The old version may be found in the <a href="https://github.com/dcgerard/updogAlpha" class="external-link"><code>updogAlpha</code> package</a>.</li>
<li>The main function is now <code><a href="../reference/flexdog.html">flexdog()</a></code>.</li>
<li>An experimental approach <code>mupdog()</code> is now live. We provide no guarantees about <code>mupdog()</code>’s performance.</li>
<li>Oracle misclassification error rates may be calculated in <code><a href="../reference/oracle_mis.html">oracle_mis()</a></code>.</li>
<li>Genotypes can be simulated using <code><a href="../reference/rgeno.html">rgeno()</a></code>.</li>
<li>Next-generation sequencing data can be simulated using <code><a href="../reference/rflexdog.html">rflexdog()</a></code>.</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

