<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="David Gerard" />


<title>Check BB Integral</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">mupdog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dcgerard/mupdog">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Check BB Integral</h1>
<h4 class="author"><em>David Gerard</em></h4>
<h4 class="date"><em>December 7, 2017</em></h4>

</div>


<div id="abstract" class="section level1">
<h1>Abstract</h1>
<p>I verify that my integral calculation is indeed correct.</p>
</div>
<div id="set-up-function" class="section level1">
<h1>Set up function</h1>
<pre class="r"><code>library(updog)
library(SuppDists)
library(gaussquad)</code></pre>
<pre><code>## Loading required package: polynom</code></pre>
<pre><code>## Loading required package: orthopolynom</code></pre>
<pre class="r"><code>get_alpha &lt;- function(mu, rho) {
  mu * (1 - rho) / rho
}

get_beta &lt;- function(mu, rho) {
  (1 - mu) * (1 - rho) / rho
}

qbetabinom &lt;- function(p, size, alpha, beta) {
  qghyper(p = p, a = -alpha, k = size, N = -alpha - beta)
}

qbetabinom_mu_rho &lt;- function(p, size, mu, rho) {
  alpha &lt;- get_alpha(mu = mu, rho = rho)
  beta  &lt;- get_beta(mu = mu, rho = rho)
  qbetabinom(p = p, size = size, alpha = alpha, beta = beta)
}

dbetabinom_hyper &lt;- function(x, size, alpha, beta, log = FALSE) {
  dghyper(x = x, a = -alpha, k = size, N = -alpha - beta, log = log)
}

dbetabinom_hyper_mu_rho &lt;- function(x, size, mu, rho, log = FALSE) {
  alpha &lt;- get_alpha(mu = mu, rho = rho)
  beta  &lt;- get_beta(mu = mu, rho = rho)
  dbetabinom_hyper(x = x, size = size, alpha = alpha, beta = beta, log = log) 
}

pbetabinom_hyper &lt;- function(q, size, alpha, beta, log.p = FALSE) {
  pghyper(q = q, a = -alpha, k = size, N = -alpha - beta, log.p = log.p)
}

pbetabinom_hyper_mu_rho &lt;- function(q, size, mu, rho, log.p = FALSE) {
  alpha &lt;- get_alpha(mu = mu, rho = rho)
  beta  &lt;- get_beta(mu = mu, rho = rho)
  pbetabinom_hyper(q = q, size = size, alpha = alpha, beta = beta, log.p = log.p)
}

assertthat::are_equal(dbetabinom_hyper(2, 8, 2, 3), updog::dbetabinom(2, 8, 2, 3))</code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code>#&#39; @param prob The allele dosage
#&#39; @param eps The sequencing error rate.
eta_fun &lt;- function(prob, eps) {
  prob * (1 - eps) + eps * (1 - prob)
}

#&#39; @param eta probability after sequencing error rate
#&#39; @param h The bias parameter
xi_fun &lt;- function(eta, h) {
  eta / (h * (1 - eta) + eta)
}

#&#39; bb function of z that we are trying to integrate
#&#39; @param x the number of A allele reads
#&#39; @param n The number of total reads
#&#39; @param z The latent variable we are integrating over
#&#39; @param eps The sequencing error rate
#&#39; @param h The bias parameter
#&#39; @param tau The od parameter of the reads
#&#39; @param alpha The allele frequency
#&#39; @param rho The inbreeding coefficient of individual i
#&#39; @param ploidy The ploidy of the species.
obj &lt;- function(x, n, z, eps, h, tau, alpha, rho, ploidy) {
  y &lt;- qbetabinom_mu_rho(p = pnorm(z), size = ploidy, mu = alpha, rho = rho)
  eta &lt;- eta_fun(prob = y / ploidy, eps = eps)  
  xi &lt;- xi_fun(eta = eta, h = h)  
  dbetabinom_hyper_mu_rho(x = x, size = n, mu = xi, rho = tau, log = TRUE)
}</code></pre>
</div>
<div id="the-supposed-integral" class="section level1">
<h1>The supposed integral</h1>
<pre class="r"><code>intobj &lt;- function(x, n, eps, h, tau, alpha, rho, ploidy, mu, sigma2) {
   pnormvec &lt;- pnorm((c(-Inf, qnorm(pbetabinom_hyper_mu_rho(q = 0:ploidy, size = ploidy, mu = alpha, rho = rho))) - mu) / sqrt(sigma2))
   wvec &lt;- pnormvec[2:length(pnormvec)] - pnormvec[1:(length(pnormvec) - 1)]
   
   eta &lt;- eta_fun(prob = 0:ploidy / ploidy, eps = eps)
   xi  &lt;- xi_fun(eta = eta, h = h)
   ldbetavec &lt;- dbetabinom_hyper_mu_rho(x = x, size = n, mu = xi, rho = tau, log = TRUE)
   
   sum(wvec * ldbetavec)
}</code></pre>
</div>
<div id="input-some-reasonable-parameter-values" class="section level1">
<h1>Input some reasonable parameter values</h1>
<pre class="r"><code>x      &lt;- 18
n      &lt;- 100
z      &lt;- -3
h      &lt;- 1
tau    &lt;- 0.01
alpha  &lt;- 1/6
rho    &lt;- 0.01
ploidy &lt;- 6
sigma2 &lt;- 1
mu     &lt;- 0
eps    &lt;- 0.01</code></pre>
<p>Compare to stochastic approximation</p>
<pre class="r"><code>library(tidyverse)</code></pre>
<pre><code>## Loading tidyverse: ggplot2
## Loading tidyverse: tibble
## Loading tidyverse: tidyr
## Loading tidyverse: readr
## Loading tidyverse: purrr
## Loading tidyverse: dplyr</code></pre>
<pre><code>## Conflicts with tidy packages ----------------------------------------------</code></pre>
<pre><code>## filter(): dplyr, stats
## lag():    dplyr, stats</code></pre>
<pre class="r"><code>itermax &lt;- 10
muvec &lt;- rnorm(itermax)
sigma2vec &lt;- rchisq(n = itermax, df = 1)

dfdat &lt;- matrix(NA, nrow = itermax, ncol = 4)
for (index in 1:itermax) {
  mu       &lt;- muvec[index]
  sigma2   &lt;- sigma2vec[index]
  zvec     &lt;- rnorm(n = 10000, mean = mu, sd = sqrt(sigma2))
  objout   &lt;- obj(x = x, n = n, z = zvec, eps = eps, h = h, tau = tau, alpha = alpha, rho = rho, ploidy = ploidy)
  mean_obj &lt;- mean(objout)
  lower    &lt;- mean_obj - 2 * sd(objout) / sqrt(length(objout))
  upper    &lt;- mean_obj + 2 * sd(objout) / sqrt(length(objout))
  theo_int &lt;- intobj(x = x, n = n, eps = eps, h = h, tau = tau, alpha = alpha, rho = rho, ploidy = ploidy, mu = mu, sigma2 = sigma2)
  dfdat[index,] &lt;- c(theo_int, mean_obj, lower, upper)
}</code></pre>
<pre class="r"><code>dfdat &lt;- as_data_frame(dfdat)
names(dfdat) &lt;- c(&quot;theo&quot;, &quot;sim&quot;, &quot;lower&quot;, &quot;upper&quot;)
ggplot(dfdat, aes(x = theo, y = sim)) +
  geom_point() +
  theme_bw() +
  geom_abline()</code></pre>
<p><img src="check_bb_integral_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 3.3.2 (2016-10-31)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.3 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] dplyr_0.7.4        purrr_0.2.2.2      readr_1.0.0       
##  [4] tidyr_0.6.1        tibble_1.3.3       ggplot2_2.2.1     
##  [7] tidyverse_1.1.1    gaussquad_1.0-2    orthopolynom_1.0-5
## [10] polynom_1.3-9      SuppDists_1.1-9.4  updog_0.1.0       
## 
## loaded via a namespace (and not attached):
##  [1] reshape2_1.4.2   haven_1.0.0      lattice_0.20-34  colorspace_1.3-2
##  [5] htmltools_0.3.6  yaml_2.1.14      rlang_0.1.2      foreign_0.8-67  
##  [9] glue_1.1.1       modelr_0.1.0     readxl_0.1.1     bindrcpp_0.2    
## [13] bindr_0.1        plyr_1.8.4       stringr_1.2.0    munsell_0.4.3   
## [17] gtable_0.2.0     rvest_0.3.2      psych_1.6.12     evaluate_0.10.1 
## [21] labeling_0.3     knitr_1.16       forcats_0.2.0    parallel_3.3.2  
## [25] broom_0.4.2      Rcpp_0.12.13     backports_1.0.5  scales_0.4.1    
## [29] jsonlite_1.5     mnormt_1.5-5     hms_0.3          digest_0.6.12   
## [33] stringi_1.1.2    grid_3.3.2       rprojroot_1.2    tools_3.3.2     
## [37] magrittr_1.5     lazyeval_0.2.0   pkgconfig_2.0.1  xml2_1.1.1      
## [41] lubridate_1.6.0  assertthat_0.2.0 rmarkdown_1.6    httr_1.2.1      
## [45] R6_2.2.2         nlme_3.1-131</code></pre>
</div>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
