<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content='Implements empirical Bayes approaches to genotype
       polyploids from next generation sequencing data while
       accounting for allele bias, overdispersion, and sequencing
       error. The main functions are flexdog() and multidog(), 
       which allow the specification
       of many different genotype distributions. Also provided are functions to
       simulate genotypes, rgeno(), and read-counts, rflexdog(), as well as
       functions to calculate oracle genotyping error rates, oracle_mis(), and
       correlation with the true genotypes, oracle_cor(). These latter two
       functions are useful for read depth calculations. Run
       browseVignettes(package = "updog") in R for example usage. See
       Gerard et al. (2018) &lt;doi:10.1534/genetics.118.301468&gt; and
       Gerard and Ferrao (2020) &lt;doi:10.1093/bioinformatics/btz852&gt; for details 
       on the implemented methods.'>
<title>Flexible Genotyping for Polyploids • updog</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Flexible Genotyping for Polyploids">
<meta property="og:description" content='Implements empirical Bayes approaches to genotype
       polyploids from next generation sequencing data while
       accounting for allele bias, overdispersion, and sequencing
       error. The main functions are flexdog() and multidog(), 
       which allow the specification
       of many different genotype distributions. Also provided are functions to
       simulate genotypes, rgeno(), and read-counts, rflexdog(), as well as
       functions to calculate oracle genotyping error rates, oracle_mis(), and
       correlation with the true genotypes, oracle_cor(). These latter two
       functions are useful for read depth calculations. Run
       browseVignettes(package = "updog") in R for example usage. See
       Gerard et al. (2018) &lt;doi:10.1534/genetics.118.301468&gt; and
       Gerard and Ferrao (2020) &lt;doi:10.1093/bioinformatics/btz852&gt; for details 
       on the implemented methods.'>
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">updog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/multidog.html">Genotyping Many SNPs with multidog()</a>
    <a class="dropdown-item" href="articles/oracle_calculations.html">Oracle Calculations</a>
    <a class="dropdown-item" href="articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    <a class="dropdown-item" href="articles/smells_like_updog.html">Example Use of Updog</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/dcgerard/updog/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="updog-">updog <a class="anchor" aria-label="anchor" href="#updog-"></a>
</h1>
</div>
<p><a href="https://github.com/dcgerard/updog/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/dcgerard/updog/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a> <a href="https://app.codecov.io/gh/dcgerard/updog" class="external-link"><img src="https://codecov.io/gh/dcgerard/updog/branch/master/graph/badge.svg?token=zjDBKZWjCs" alt="codecov"></a> <a href="https://www.gnu.org/licenses/gpl-3.0" class="external-link"><img src="https://img.shields.io/badge/License-GPL%20v3-blue.svg" alt="License: GPL v3"></a> <a href="https://cran.r-project.org/package=updog" class="external-link"><img src="https://www.r-pkg.org/badges/version/updog" alt="CRAN_Status_Badge"></a> <a href="https://cran.r-project.org/package=updog" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/grand-total/updog"></a></p>
<p>Updog provides a suite of methods for genotyping polyploids from next-generation sequencing (NGS) data. It does this while accounting for many common features of NGS data: allele bias, overdispersion, and sequencing error. It is named updog for “Using Parental Data for Offspring Genotyping” because we originally developed the method for full-sib populations, but it works now for more general populations. The method is described in detail Gerard et. al. (2018) &lt;<a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">doi:10.1534/genetics.118.301468</a>&gt;. Additional details concerning prior specification are described in Gerard and Ferrão (2020) &lt;<a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link">doi:10.1093/bioinformatics/btz852</a>&gt;.</p>
<p>The main functions are <code><a href="reference/flexdog.html">flexdog()</a></code> and <code><a href="reference/multidog.html">multidog()</a></code>, which provide many options for the distribution of the genotypes in your sample. A novel genotype distribution is included in the class of proportional normal distributions (<code>model = "norm"</code>). This is the default prior distribution because it is the most robust to varying genotype distributions, but feel free to use more specialized priors if you have more information on the data.</p>
<p>Also provided are:</p>
<ul>
<li>
<code><a href="reference/filter_snp.html">filter_snp()</a></code>: filter out SNPs based on the output of <code><a href="reference/multidog.html">multidog()</a></code>.</li>
<li>
<code><a href="reference/format_multidog.html">format_multidog()</a></code>: format the output of <code><a href="reference/multidog.html">multidog()</a></code> in terms of a multidimensional array.</li>
<li>Plot methods. Both <code><a href="reference/flexdog.html">flexdog()</a></code> and <code><a href="reference/multidog.html">multidog()</a></code> have plot methods. See the help files of <code><a href="reference/plot.flexdog.html">plot.flexdog()</a></code> and <code><a href="reference/plot.multidog.html">plot.multidog()</a></code> for details.</li>
<li>Functions to simulate genotypes (<code><a href="reference/rgeno.html">rgeno()</a></code>) and read-counts (<code><a href="reference/rflexdog.html">rflexdog()</a></code>). These support all of the models available in <code><a href="reference/flexdog.html">flexdog()</a></code>.</li>
<li>Functions to evaluate oracle genotyping performance: <code><a href="reference/oracle_joint.html">oracle_joint()</a></code>, <code><a href="reference/oracle_mis.html">oracle_mis()</a></code>, <code><a href="reference/oracle_mis_vec.html">oracle_mis_vec()</a></code>, and <code><a href="reference/oracle_cor.html">oracle_cor()</a></code>. We mean “oracle” in the sense that we assume that the entire data generation process is known (i.e. the genotype distribution, sequencing error rate, allele bias, and overdispersion are all known). These are good approximations when there are a lot of individuals (but not necessarily large read-depth).</li>
</ul>
<p>The original <code>updog</code> package is now named <code>updogAlpha</code> and may be found <a href="https://github.com/dcgerard/updogAlpha" class="external-link">here</a>.</p>
<p>See also <a href="https://github.com/pblischak/polyploid-genotyping" class="external-link">ebg</a>, <a href="https://cran.r-project.org/package=fitPoly" class="external-link">fitPoly</a>, and <a href="https://cran.r-project.org/package=polyRAD" class="external-link">polyRAD</a>. Our best “competitor” is probably <a href="https://cran.r-project.org/package=fitPoly" class="external-link">fitPoly</a>, though <a href="https://cran.r-project.org/package=polyRAD" class="external-link">polyRAD</a> has some nice ideas for utilizing population structure and linkage disequilibrium.</p>
<p>See <a href="https://github.com/dcgerard/updog/blob/master/NEWS.md" class="external-link">NEWS</a> for the latest updates on the package.</p>
<div class="section level2">
<h2 id="vignettes">Vignettes<a class="anchor" aria-label="anchor" href="#vignettes"></a>
</h2>
<p>I’ve included many vignettes in <code>updog</code>, which you can access online <a href="https://dcgerard.github.io/updog/" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="bug-reports">Bug Reports<a class="anchor" aria-label="anchor" href="#bug-reports"></a>
</h2>
<p>If you find a bug or want an enhancement, please submit an issue <a href="https://github.com/dcgerard/updog/issues" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install updog from <a href="https://cran.r-project.org/package=updog" class="external-link">CRAN</a> in the usual way:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"updog"</span><span class="op">)</span></span></code></pre></div>
<p>You can install the current (unstable) version of updog from GitHub with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"dcgerard/updog"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="how-to-cite">How to Cite<a class="anchor" aria-label="anchor" href="#how-to-cite"></a>
</h2>
<p>Please cite</p>
<blockquote>
<p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. doi: <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">10.1534/genetics.118.301468</a>.</p>
</blockquote>
<p>Or, using BibTex:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode tex"><code class="sourceCode latex"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>@article {gerard2018genotyping,</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>    author = {Gerard, David and Ferr{<span class="fu">\~</span>a}o, Lu{<span class="fu">\'</span>i}s Felipe Ventorim and Garcia, Antonio Augusto Franco and Stephens, Matthew},</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>    title = {Genotyping Polyploids from Messy Sequencing Data},</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>    volume = {210},</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>    number = {3},</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>    pages = {789--807},</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>    year = {2018},</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>    doi = {10.1534/genetics.118.301468},</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>    publisher = {Genetics},</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>    issn = {0016-6731},</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>    URL = {https://doi.org/10.1534/genetics.118.301468},</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>    journal = {Genetics}</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>}</span></code></pre></div>
<p>If you are using the proportional normal prior class (<code>model = "norm"</code>), which is also the default prior, then please also cite:</p>
<blockquote>
<p>Gerard D, Ferrão L (2020). “Priors for Genotyping Polyploids.” <em>Bioinformatics</em>, 36(6), 1795-1800. ISSN 1367-4803, doi: <a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link">10.1093/bioinformatics/btz852</a>.</p>
</blockquote>
<p>Or, using BibTex:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode tex"><code class="sourceCode latex"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>@article{gerard2020priors,</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>    title = {Priors for Genotyping Polyploids},</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>    year = {2020},</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>    journal = {Bioinformatics},</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>    publisher = {Oxford University Press},</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>    volume = {36},</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>    number = {6},</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>    pages = {1795--1800},</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>    issn = {1367-4803},</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>    doi = {10.1093/bioinformatics/btz852},</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>    author = {David Gerard and Lu{<span class="fu">\'</span>i}s Felipe Ventorim Ferr{<span class="fu">\~</span>a}o},</span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>  }</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="code-of-conduct">Code of Conduct<a class="anchor" aria-label="anchor" href="#code-of-conduct"></a>
</h2>
<p>Please note that this project is released with a <a href="https://github.com/dcgerard/updog/blob/master/CONDUCT.md" class="external-link">Contributor Code of Conduct</a>. By participating in this project you agree to abide by its terms.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=updog" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/dcgerard/updog/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/dcgerard/updog/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing updog</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>David Gerard <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9450-5023" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
