<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="updog">
<title>Genotyping Many SNPs with multidog() • updog</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Genotyping Many SNPs with multidog()">
<meta property="og:description" content="updog">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">updog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a>
    <a class="dropdown-item" href="../articles/oracle_calculations.html">Oracle Calculations</a>
    <a class="dropdown-item" href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    <a class="dropdown-item" href="../articles/smells_like_updog.html">Example Use of Updog</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/dcgerard/updog/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="multidog_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Genotyping Many SNPs with multidog()</h1>
                        <h4 data-toc-skip class="author">David Gerard</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/vignettes/multidog.Rmd" class="external-link"><code>vignettes/multidog.Rmd</code></a></small>
      <div class="d-none name"><code>multidog.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<p><code><a href="../reference/multidog.html">multidog()</a></code> provides support for genotyping many SNPs by iterating <code><a href="../reference/flexdog.html">flexdog()</a></code> over the SNPs. Support is provided for parallel computing through the <a href="https://cran.r-project.org/package=future" class="external-link"><code>future</code></a> package. The genotyping method is described in Gerard et al. (2018) and Gerard and Ferrão (2020).</p>
</div>
<div class="section level2">
<h2 id="fit-multidog">Fit <code>multidog()</code><a class="anchor" aria-label="anchor" href="#fit-multidog"></a>
</h2>
<p>Let’s load <code>updog</code>, <code>future</code>, and the data from Uitdewilligen et al. (2013).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dcgerard/updog/" class="external-link">updog</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"uitdewilligen"</span><span class="op">)</span></span></code></pre></div>
<p><code>uitdewilligen$refmat</code> is a matrix of reference counts while <code>uitdewilligen$sizemat</code> is a matrix of total read counts. In these data, the rows index the individuals and the columns index the loci. But for insertion into <code><a href="../reference/multidog.html">multidog()</a></code> we need it the other way around (individuals in the columns and loci in the rows). So we will transpose these matrices.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">refmat</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">uitdewilligen</span><span class="op">$</span><span class="va">refmat</span><span class="op">)</span></span>
<span><span class="va">sizemat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">uitdewilligen</span><span class="op">$</span><span class="va">sizemat</span><span class="op">)</span></span>
<span><span class="va">ploidy</span>  <span class="op">&lt;-</span> <span class="va">uitdewilligen</span><span class="op">$</span><span class="va">ploidy</span></span></code></pre></div>
<p><code>sizemat</code> and <code>refmat</code> should have the same row and column names. These names identify the loci and the individuals.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sizemat</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">refmat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; character(0)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sizemat</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">refmat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; character(0)</span></span></code></pre></div>
<p>If we want to do parallel computing, we should check that we have the proper number of cores:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; system </span></span>
<span><span class="co">#&gt;     16</span></span></code></pre></div>
<p>Now let’s run <code><a href="../reference/multidog.html">multidog()</a></code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mout</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/multidog.html">multidog</a></span><span class="op">(</span>refmat <span class="op">=</span> <span class="va">refmat</span>, </span>
<span>                 sizemat <span class="op">=</span> <span class="va">sizemat</span>, </span>
<span>                 ploidy <span class="op">=</span> <span class="va">ploidy</span>, </span>
<span>                 model <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>                 nc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt;     |                                   *.#,%    </span></span>
<span><span class="co">#&gt;    |||                                 *******/  </span></span>
<span><span class="co">#&gt;  |||||||    (**..#**.                  */   **/  </span></span>
<span><span class="co">#&gt; |||||||||    */****************************/*%   </span></span>
<span><span class="co">#&gt;    |||    &amp;****..,*.************************/    </span></span>
<span><span class="co">#&gt;    |||     (....,,,*,...****%********/(******    </span></span>
<span><span class="co">#&gt;    |||                ,,****%////,,,,./.****/    </span></span>
<span><span class="co">#&gt;    |||                  /**//         .*///....  </span></span>
<span><span class="co">#&gt;    |||                  .*/*/%#         .,/   ., </span></span>
<span><span class="co">#&gt;    |||               , **/   #%         .*    .. </span></span>
<span><span class="co">#&gt;    |||                               ,,,*        </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Working on it...done!</span></span></code></pre></div>
<p>By default, parallelization is run using</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="fu">future</span><span class="fu">::</span><span class="va"><a href="https://future.futureverse.org/reference/multisession.html" class="external-link">multisession</a></span>, workers <span class="op">=</span> <span class="va">nc</span><span class="op">)</span></span></code></pre></div>
<p>if <code>nc</code> is greater than 1. You can choose your own evaluation strategy by running <code><a href="https://future.futureverse.org/reference/plan.html" class="external-link">future::plan()</a></code> prior to running <code><a href="../reference/multidog.html">multidog()</a></code>, and then setting <code>nc = NA</code>. This should be particularly useful in higher performance computing environments that use schedulers, where you can control the evaluation strategy through the <a href="https://cran.r-project.org/package=future.batchtools" class="external-link"><code>future.batchtools</code></a> package. For example, the following will run <code><a href="../reference/multidog.html">multidog()</a></code> using forked R processes:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="fu">future</span><span class="fu">::</span><span class="va"><a href="https://future.futureverse.org/reference/multicore.html" class="external-link">multicore</a></span>, workers <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">mout</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/multidog.html">multidog</a></span><span class="op">(</span>refmat <span class="op">=</span> <span class="va">refmat</span>, </span>
<span>                 sizemat <span class="op">=</span> <span class="va">sizemat</span>, </span>
<span>                 ploidy <span class="op">=</span> <span class="va">ploidy</span>, </span>
<span>                 model <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>                 nc <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Shut down parallel workers</span></span>
<span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="fu">future</span><span class="fu">::</span><span class="va"><a href="https://future.futureverse.org/reference/sequential.html" class="external-link">sequential</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="multidog-output">
<code>multidog()</code> Output<a class="anchor" aria-label="anchor" href="#multidog-output"></a>
</h2>
<p>There is a plot method for the output of <code><a href="../reference/multidog.html">multidog()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">mout</span>, indices <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span></code></pre></div>
<p><img src="multidog_files/figure-html/unnamed-chunk-8-1.png" width="432"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span></code></pre>
<p><img src="multidog_files/figure-html/unnamed-chunk-8-2.png" width="432"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span></code></pre>
<p><img src="multidog_files/figure-html/unnamed-chunk-8-3.png" width="432"></p>
<p>The output of multidog contains two data frame. The first contains properties of the SNPs, such as estimated allele bias and estimated sequencing error rate.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">mout</span><span class="op">$</span><span class="va">snpdf</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    100 obs. of  20 variables:</span></span>
<span><span class="co">#&gt;  $ snp     : chr  "PotVar0089524" "PotVar0052647" "PotVar0120897" "PotVar0066020" ...</span></span>
<span><span class="co">#&gt;  $ bias    : num  0.519 1.026 0.929 1.221 0.847 ...</span></span>
<span><span class="co">#&gt;  $ seq     : num  0.00485 0.00221 0.002 0.0039 0.00206 ...</span></span>
<span><span class="co">#&gt;  $ od      : num  0.00304 0.00295 0.00337 0.00275 0.00335 ...</span></span>
<span><span class="co">#&gt;  $ prop_mis: num  0.004926 0.002274 0.000626 0.002718 0.003 ...</span></span>
<span><span class="co">#&gt;  $ num_iter: num  6 3 3 5 7 7 4 8 8 4 ...</span></span>
<span><span class="co">#&gt;  $ llike   : num  -14.7 -25.3 -10.4 -22.7 -32 ...</span></span>
<span><span class="co">#&gt;  $ ploidy  : num  4 4 4 4 4 4 4 4 4 4 ...</span></span>
<span><span class="co">#&gt;  $ model   : chr  "norm" "norm" "norm" "norm" ...</span></span>
<span><span class="co">#&gt;  $ p1ref   : num  NA NA NA NA NA NA NA NA NA NA ...</span></span>
<span><span class="co">#&gt;  $ p1size  : num  NA NA NA NA NA NA NA NA NA NA ...</span></span>
<span><span class="co">#&gt;  $ p2ref   : num  NA NA NA NA NA NA NA NA NA NA ...</span></span>
<span><span class="co">#&gt;  $ p2size  : num  NA NA NA NA NA NA NA NA NA NA ...</span></span>
<span><span class="co">#&gt;  $ Pr_0    : num  0.000279 0.248211 0.66369 0.015803 0.08409 ...</span></span>
<span><span class="co">#&gt;  $ Pr_1    : num  0.00707 0.45067 0.26892 0.06938 0.20154 ...</span></span>
<span><span class="co">#&gt;  $ Pr_2    : num  0.0745 0.2542 0.0597 0.1931 0.2968 ...</span></span>
<span><span class="co">#&gt;  $ Pr_3    : num  0.32604 0.04452 0.00725 0.34069 0.26844 ...</span></span>
<span><span class="co">#&gt;  $ Pr_4    : num  0.592065 0.002423 0.000482 0.381024 0.149179 ...</span></span>
<span><span class="co">#&gt;  $ mu      : num  4.18 1.01 -1 3.75 2.29 ...</span></span>
<span><span class="co">#&gt;  $ sigma   : num  1.067 0.925 1.289 1.481 1.433 ...</span></span></code></pre></div>
<p>The second data frame contains properties of each individual at each SNP, such as the estimated genotypes (<code>geno</code>) and the posterior probability of being genotyping correctly (<code>maxpostprob</code>).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">mout</span><span class="op">$</span><span class="va">inddf</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    1000 obs. of  17 variables:</span></span>
<span><span class="co">#&gt;  $ snp        : chr  "PotVar0089524" "PotVar0089524" "PotVar0089524" "PotVar0089524" ...</span></span>
<span><span class="co">#&gt;  $ ind        : chr  "P5PEM08" "P3PEM05" "P2PEM10" "P7PEM09" ...</span></span>
<span><span class="co">#&gt;  $ ref        : num  122 113 86 80 69 85 130 228 60 211 ...</span></span>
<span><span class="co">#&gt;  $ size       : num  142 143 96 80 69 86 130 228 86 212 ...</span></span>
<span><span class="co">#&gt;  $ geno       : num  3 3 3 4 4 4 4 4 2 4 ...</span></span>
<span><span class="co">#&gt;  $ postmean   : num  3 2.99 3 4 4 ...</span></span>
<span><span class="co">#&gt;  $ maxpostprob: num  1 0.988 1 1 1 ...</span></span>
<span><span class="co">#&gt;  $ Pr_0       : num  3.74e-90 1.03e-78 2.21e-77 1.06e-86 8.21e-79 ...</span></span>
<span><span class="co">#&gt;  $ Pr_1       : num  7.97e-23 3.86e-16 2.61e-20 6.80e-30 1.21e-26 ...</span></span>
<span><span class="co">#&gt;  $ Pr_2       : num  4.94e-06 1.17e-02 3.27e-06 2.82e-14 1.01e-12 ...</span></span>
<span><span class="co">#&gt;  $ Pr_3       : num  1.00 9.88e-01 1.00 6.74e-06 2.75e-05 ...</span></span>
<span><span class="co">#&gt;  $ Pr_4       : num  1.45e-10 1.14e-15 3.56e-06 1.00 1.00 ...</span></span>
<span><span class="co">#&gt;  $ logL_0     : num  -201 -176 -172 -190 -172 ...</span></span>
<span><span class="co">#&gt;  $ logL_1     : num  -49.6 -35.6 -44 -62.9 -55.4 ...</span></span>
<span><span class="co">#&gt;  $ logL_2     : num  -13.27 -6.95 -13.93 -29.29 -25.69 ...</span></span>
<span><span class="co">#&gt;  $ logL_3     : num  -2.55 -4 -2.79 -11.49 -10.06 ...</span></span>
<span><span class="co">#&gt;  $ logL_4     : num  -25.804 -38.999 -15.935 -0.181 -0.158 ...</span></span></code></pre></div>
<p>You can obtain the columns in <code>inddf</code> in matrix form with <code><a href="../reference/format_multidog.html">format_multidog()</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_multidog.html">format_multidog</a></span><span class="op">(</span><span class="va">mout</span>, varname <span class="op">=</span> <span class="st">"geno"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">genomat</span><span class="op">)</span></span>
<span><span class="co">#&gt;               P1PEM10 P2PEM05 P2PEM10 P3PEM05 P4PEM01 P4PEM09 P5PEM04 P5PEM08</span></span>
<span><span class="co">#&gt; PotVar0089524       4       4       3       3       4       4       4       3</span></span>
<span><span class="co">#&gt; PotVar0052647       3       1       0       1       1       2       0       1</span></span>
<span><span class="co">#&gt; PotVar0120897       0       0       0       0       0       0       0       1</span></span>
<span><span class="co">#&gt; PotVar0066020       3       2       3       4       4       3       1       4</span></span>
<span><span class="co">#&gt; PotVar0003381       3       1       2       0       2       3       3       1</span></span>
<span><span class="co">#&gt; PotVar0131622       2       4       1       2       2       3       4       3</span></span>
<span><span class="co">#&gt;               P6PEM11 P7PEM09</span></span>
<span><span class="co">#&gt; PotVar0089524       2       4</span></span>
<span><span class="co">#&gt; PotVar0052647       1       1</span></span>
<span><span class="co">#&gt; PotVar0120897       2       1</span></span>
<span><span class="co">#&gt; PotVar0066020       4       2</span></span>
<span><span class="co">#&gt; PotVar0003381       4       3</span></span>
<span><span class="co">#&gt; PotVar0131622       3       3</span></span></code></pre></div>
<p>To filter SNPs based on quality metrics (bias, sequencing error rate, overdispersion, etc), you can use <code><a href="../reference/filter_snp.html">filter_snp()</a></code>, which uses the same non-standard evaluation you are used to from <code><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">dplyr::filter()</a></code>. That is, you can define predicates in terms of the variable names in the <code>snpdf</code> data frame from the output of <code>mupdog()</code>. It then keeps rows in both <code>snpdf</code> and <code>inddf</code> where the predicate for a SNP evaluates to <code>TRUE</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mout</span><span class="op">$</span><span class="va">snpdf</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 100  20</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mout</span><span class="op">$</span><span class="va">inddf</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1000   17</span></span>
<span><span class="va">mout_cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_snp.html">filter_snp</a></span><span class="op">(</span><span class="va">mout</span>, <span class="va">prop_mis</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">bias</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">bias</span> <span class="op">&lt;</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mout_cleaned</span><span class="op">$</span><span class="va">snpdf</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 97 20</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mout_cleaned</span><span class="op">$</span><span class="va">inddf</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 970  17</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Gerard, David, and Luís Felipe Ventorim Ferrão. “Priors for genotyping polyploids.” <em>Bioinformatics</em> 36, no. 6 (2020): 1795-1800. <a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link uri">https://doi.org/10.1093/bioinformatics/btz852</a>.</p>
<p>Gerard, David, Luís Felipe Ventorim Ferrão, Antonio Augusto Franco Garcia, and Matthew Stephens. 2018. “Genotyping Polyploids from Messy Sequencing Data.” <em>Genetics</em> 210 (3). Genetics: 789–807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link uri">https://doi.org/10.1534/genetics.118.301468</a>.</p>
<p>Uitdewilligen, Anne-Marie A. AND D’hoop, Jan G. A. M. L. AND Wolters. 2013. “A Next-Generation Sequencing Method for Genotyping-by-Sequencing of Highly Heterozygous Autotetraploid Potato.” <em>PLOS ONE</em> 8 (5). Public Library of Science: 1–14. <a href="https://doi.org/10.1371/journal.pone.0062355" class="external-link uri">https://doi.org/10.1371/journal.pone.0062355</a>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
