<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Oracle Calculations • updog</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Oracle Calculations">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">updog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a></li>
    <li><a class="dropdown-item" href="../articles/oracle_calculations.html">Oracle Calculations</a></li>
    <li><a class="dropdown-item" href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a></li>
    <li><a class="dropdown-item" href="../articles/smells_like_updog.html">Example Use of Updog</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/dcgerard/updog/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Oracle Calculations</h1>
                        <h4 data-toc-skip class="author">David
Gerard</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/vignettes/oracle_calculations.Rmd" class="external-link"><code>vignettes/oracle_calculations.Rmd</code></a></small>
      <div class="d-none name"><code>oracle_calculations.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<p>We provide some example usage of the oracle calculations available in
<code>updog</code>. These are particularly useful for read-depth
determination. These calculations are described in detail in Gerard et
al. (2018).</p>
</div>
<div class="section level2">
<h2 id="controlling-misclassification-error">Controlling Misclassification Error<a class="anchor" aria-label="anchor" href="#controlling-misclassification-error"></a>
</h2>
<p>Suppose we have a sample of tetraploid individuals derived from an S1
cross (a single generation of selfing). Using domain expertise (either
from previous studies or a pilot analysis), we’ve determined that our
sequencing technology will produce relatively clean data. That is, the
sequencing error rate will not be too large (say, ~0.001), the bias will
be moderate (say, ~0.7 at the most extreme), and the majority of SNPs
will have reasonable levels of overdispersion (say, less than 0.01). We
want to know how deep we need to sequence.</p>
<p>Using <code>oracle_mis</code>, we can see how deep we need to
sequence under the worst-case scenario we want to control (sequencing
error rate = 0.001, bias = 0.7, overdispersion = 0.01) in order to
obtain a misclassification error rate of at most, say, 0.05.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bias</span>   <span class="op">&lt;-</span> <span class="fl">0.7</span></span>
<span><span class="va">od</span>     <span class="op">&lt;-</span> <span class="fl">0.01</span></span>
<span><span class="va">seq</span>    <span class="op">&lt;-</span> <span class="fl">0.001</span></span>
<span><span class="va">maxerr</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span></code></pre></div>
<p>Before we do this, we also need the distribution of the offspring
genotypes. We can get this distribution assuming various parental
genotypes using the <code>get_q_array</code> function. Typically, error
rates will be larger when the allele-frequency is closer to 0.5. So
we’ll start in the worst-case scenario of assuming that the parent has 2
copies of the reference allele.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dcgerard/updog/" class="external-link">updog</a></span><span class="op">)</span></span>
<span><span class="va">ploidy</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span><span class="va">pgeno</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="va">gene_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_q_array.html">get_q_array</a></span><span class="op">(</span>ploidy <span class="op">=</span> <span class="va">ploidy</span><span class="op">)</span><span class="op">[</span><span class="va">pgeno</span> <span class="op">+</span> <span class="fl">1</span>, <span class="va">pgeno</span> <span class="op">+</span> <span class="fl">1</span>, <span class="op">]</span></span></code></pre></div>
<p>This is what the genotype distribution for the offspring looks
like:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">distdf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="va">ploidy</span>, y <span class="op">=</span> <span class="fl">0</span>, yend <span class="op">=</span> <span class="va">gene_dist</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">distdf</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, xend <span class="op">=</span> <span class="va">x</span>, yend <span class="op">=</span> <span class="va">yend</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span>lineend <span class="op">=</span> <span class="st">"round"</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Allele Dosage"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Probability"</span><span class="op">)</span></span></code></pre></div>
<p><img src="oracle_calculations_files/figure-html/unnamed-chunk-3-1.png" width="432"></p>
<p>Now, we are ready to iterate through read-depth’s until we reach one
with an error rate less than 0.05.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">err</span>    <span class="op">&lt;-</span> <span class="cn">Inf</span></span>
<span><span class="va">depth</span>  <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="kw">while</span><span class="op">(</span><span class="va">err</span> <span class="op">&gt;</span> <span class="va">maxerr</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">depth</span> <span class="op">&lt;-</span> <span class="va">depth</span> <span class="op">+</span> <span class="fl">1</span></span>
<span>  <span class="va">err</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oracle_mis.html">oracle_mis</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">depth</span>,</span>
<span>                    ploidy <span class="op">=</span> <span class="va">ploidy</span>,</span>
<span>                    seq <span class="op">=</span> <span class="va">seq</span>,</span>
<span>                    bias <span class="op">=</span> <span class="va">bias</span>,</span>
<span>                    od <span class="op">=</span> <span class="va">od</span>,</span>
<span>                    dist <span class="op">=</span> <span class="va">gene_dist</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">depth</span></span>
<span><span class="co">#&gt; [1] 90</span></span></code></pre></div>
<p>Looks like we need a depth of 90 in order to get a misclassification
error rate under 0.05.</p>
<p>Note that <code>oracle_mis</code> returns the <strong>best
misclassification error rate possible</strong> under these conditions
(<code>ploidy</code> = 4, <code>bias</code> = 0.7, <code>seq</code> =
0.001, <code>od</code> = 0.01, and <code>pgeno</code> = 2). In your
actual analysis, you will have a worse misclassification error rate than
that returned by <code>oracle_mis</code>. However, if you have a lot of
individuals in your sample, then this will act as a reasonable
approximation to the error rate. In general though, you should sequence
a little deeper than suggested by <code>oracle_mis</code>.</p>
</div>
<div class="section level2">
<h2 id="visualizing-the-joint-distribution">Visualizing the Joint Distribution<a class="anchor" aria-label="anchor" href="#visualizing-the-joint-distribution"></a>
</h2>
<p>Suppose we only have a budget to sequence to a depth of 30. Then what
errors can we expect? We can use <code>oracle_joint</code> and
<code>oracle_plot</code> to visualize the errors we can expect.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">depth</span> <span class="op">&lt;-</span> <span class="fl">30</span></span>
<span><span class="va">jd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oracle_joint.html">oracle_joint</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">depth</span>,</span>
<span>                   ploidy <span class="op">=</span> <span class="va">ploidy</span>,</span>
<span>                   seq <span class="op">=</span> <span class="va">seq</span>,</span>
<span>                   bias <span class="op">=</span> <span class="va">bias</span>,</span>
<span>                   od <span class="op">=</span> <span class="va">od</span>,</span>
<span>                   dist <span class="op">=</span> <span class="va">gene_dist</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/oracle_plot.html">oracle_plot</a></span><span class="op">(</span><span class="va">jd</span><span class="op">)</span></span></code></pre></div>
<p><img src="oracle_calculations_files/figure-html/unnamed-chunk-5-1.png" width="432"></p>
<p>Most of the errors will be mistakes between genotypes 2/3 and
mistakes between genotypes 1/2.</p>
<p>Even though the misclassification error rate is pretty high (0.14),
the correlation of the oracle estimator with the true genotype is pretty
reasonable (0.89). You can obtain this using the <code>oracle_cor</code>
function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ocorr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oracle_cor.html">oracle_cor</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">depth</span>,</span>
<span>                    ploidy <span class="op">=</span> <span class="va">ploidy</span>,</span>
<span>                    seq <span class="op">=</span> <span class="va">seq</span>,</span>
<span>                    bias <span class="op">=</span> <span class="va">bias</span>,</span>
<span>                    od <span class="op">=</span> <span class="va">od</span>,</span>
<span>                    dist <span class="op">=</span> <span class="va">gene_dist</span><span class="op">)</span></span>
<span><span class="va">ocorr</span></span>
<span><span class="co">#&gt; [1] 0.8935101</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Gerard, David, Luís Felipe Ventorim Ferrão, Antonio Augusto Franco
Garcia, and Matthew Stephens. 2018. “Genotyping Polyploids from Messy
Sequencing Data.” <em>Genetics</em> 210 (3). Genetics: 789–807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link uri">https://doi.org/10.1534/genetics.118.301468</a>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
