<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Simulate Next-Generation Sequencing Data • updog</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate Next-Generation Sequencing Data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">updog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a></li>
    <li><a class="dropdown-item" href="../articles/oracle_calculations.html">Oracle Calculations</a></li>
    <li><a class="dropdown-item" href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a></li>
    <li><a class="dropdown-item" href="../articles/smells_like_updog.html">Example Use of Updog</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/dcgerard/updog/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Simulate Next-Generation Sequencing Data</h1>
                        <h4 data-toc-skip class="author">David
Gerard</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/vignettes/simulate_ngs.Rmd" class="external-link"><code>vignettes/simulate_ngs.Rmd</code></a></small>
      <div class="d-none name"><code>simulate_ngs.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<p>We demonstrate how to simulate NGS data under various genotype
distributions, then fit these data using <code>flexdog</code>. The
genotyping methods are described in Gerard et al. (2018).</p>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p>Let’s suppose that we have 100 hexaploid individuals, with varying
levels of read-depth.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dcgerard/updog/" class="external-link">updog</a></span><span class="op">)</span></span>
<span><span class="va">nind</span>    <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">ploidy</span>  <span class="op">&lt;-</span> <span class="fl">6</span></span>
<span><span class="va">sizevec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n   <span class="op">=</span> <span class="va">nind</span>, </span>
<span>                              min <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                              max <span class="op">=</span> <span class="fl">200</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can simulate their read-counts under various genotype
distributions, allele biases, overdispersions, and sequencing error
rates using the <code>rgeno</code> and <code>rflexdog</code>
functions.</p>
<div class="section level3">
<h3 id="f1-population">F1 Population<a class="anchor" aria-label="anchor" href="#f1-population"></a>
</h3>
<p>Suppose these individuals are all siblings where the first parent has
4 copies of the reference allele and the second parent has 5 copies of
the reference allele. Then the following code, using <code>rgeno</code>,
will simulate the individuals’ genotypes.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">true_geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rgeno.html">rgeno</a></span><span class="op">(</span>n      <span class="op">=</span> <span class="va">nind</span>, </span>
<span>                   ploidy <span class="op">=</span> <span class="va">ploidy</span>,</span>
<span>                   model  <span class="op">=</span> <span class="st">"f1"</span>, </span>
<span>                   p1geno <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                   p2geno <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>Once we have their genotypes, we can simulate their read-counts using
<code>rflexdog</code>. Let’s suppose that there is a moderate level of
allelic bias (0.7) and a small level of overdispersion (0.005).
Generally, in the real data that I’ve seen, the bias will range between
0.5 and 2 and the overdispersion will range between 0 and 0.02, with
only a few extremely overdispersed SNPs above 0.02.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">refvec</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rflexdog.html">rflexdog</a></span><span class="op">(</span>sizevec <span class="op">=</span> <span class="va">sizevec</span>, </span>
<span>                   geno    <span class="op">=</span> <span class="va">true_geno</span>, </span>
<span>                   ploidy  <span class="op">=</span> <span class="va">ploidy</span>, </span>
<span>                   seq     <span class="op">=</span> <span class="fl">0.001</span>, </span>
<span>                   bias    <span class="op">=</span> <span class="fl">0.7</span>, </span>
<span>                   od      <span class="op">=</span> <span class="fl">0.005</span><span class="op">)</span></span></code></pre></div>
<p>When we plot the data, it looks realistic</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_geno.html">plot_geno</a></span><span class="op">(</span>refvec  <span class="op">=</span> <span class="va">refvec</span>, </span>
<span>          sizevec <span class="op">=</span> <span class="va">sizevec</span>, </span>
<span>          ploidy  <span class="op">=</span> <span class="va">ploidy</span>, </span>
<span>          bias    <span class="op">=</span> <span class="fl">0.7</span>, </span>
<span>          seq     <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>          geno    <span class="op">=</span> <span class="va">true_geno</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulate_ngs_files/figure-html/unnamed-chunk-4-1.png" width="432"></p>
<p>We can test <code>flexdog</code> on these data</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fout</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flexdog.html">flexdog</a></span><span class="op">(</span>refvec  <span class="op">=</span> <span class="va">refvec</span>, </span>
<span>                sizevec <span class="op">=</span> <span class="va">sizevec</span>, </span>
<span>                ploidy  <span class="op">=</span> <span class="va">ploidy</span>, </span>
<span>                model   <span class="op">=</span> <span class="st">"f1"</span><span class="op">)</span></span>
<span><span class="co">#&gt;          Fit: 1 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 0.3678794 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -364.3851 </span></span>
<span><span class="co">#&gt; Keeping new fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          Fit: 2 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 0.6065307 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -364.3851 </span></span>
<span><span class="co">#&gt; Keeping old fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          Fit: 3 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 1 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -364.385 </span></span>
<span><span class="co">#&gt; Keeping new fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          Fit: 4 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 1.648721 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -381.6985 </span></span>
<span><span class="co">#&gt; Keeping old fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          Fit: 5 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 2.718282 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -412.5899 </span></span>
<span><span class="co">#&gt; Keeping old fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Done!</span></span></code></pre></div>
<p><code>flexdog</code> gives us reasonable genotyping, and it
accurately estimates the proportion of individuals mis-genotyped.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fout</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulate_ngs_files/figure-html/unnamed-chunk-6-1.png" width="432"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Estimated proportion misgenotyped</span></span>
<span><span class="va">fout</span><span class="op">$</span><span class="va">prop_mis</span></span>
<span><span class="co">#&gt; [1] 0.06990052</span></span>
<span></span>
<span><span class="co">## Actual proportion misgenotyped</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">fout</span><span class="op">$</span><span class="va">geno</span> <span class="op">!=</span> <span class="va">true_geno</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.05</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="hwe-population">HWE Population<a class="anchor" aria-label="anchor" href="#hwe-population"></a>
</h3>
<p>Now run the same simulations assuming the individuals are in
Hardy-Weinberg population with an allele frequency of 0.75.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">true_geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rgeno.html">rgeno</a></span><span class="op">(</span>n      <span class="op">=</span> <span class="va">nind</span>,</span>
<span>                   ploidy <span class="op">=</span> <span class="va">ploidy</span>,</span>
<span>                   model  <span class="op">=</span> <span class="st">"hw"</span>, </span>
<span>                   allele_freq <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span>
<span><span class="va">refvec</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rflexdog.html">rflexdog</a></span><span class="op">(</span>sizevec <span class="op">=</span> <span class="va">sizevec</span>, </span>
<span>                   geno    <span class="op">=</span> <span class="va">true_geno</span>, </span>
<span>                   ploidy  <span class="op">=</span> <span class="va">ploidy</span>, </span>
<span>                   seq     <span class="op">=</span> <span class="fl">0.001</span>, </span>
<span>                   bias    <span class="op">=</span> <span class="fl">0.7</span>, </span>
<span>                   od      <span class="op">=</span> <span class="fl">0.005</span><span class="op">)</span></span>
<span><span class="va">fout</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flexdog.html">flexdog</a></span><span class="op">(</span>refvec  <span class="op">=</span> <span class="va">refvec</span>, </span>
<span>                sizevec <span class="op">=</span> <span class="va">sizevec</span>, </span>
<span>                ploidy  <span class="op">=</span> <span class="va">ploidy</span>, </span>
<span>                model   <span class="op">=</span> <span class="st">"hw"</span><span class="op">)</span></span>
<span><span class="co">#&gt;          Fit: 1 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 0.3678794 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -378.2634 </span></span>
<span><span class="co">#&gt; Keeping new fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          Fit: 2 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 0.6065307 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -378.2633 </span></span>
<span><span class="co">#&gt; Keeping new fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          Fit: 3 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 1 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -378.2633 </span></span>
<span><span class="co">#&gt; Keeping old fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          Fit: 4 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 1.648721 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -378.2633 </span></span>
<span><span class="co">#&gt; Keeping old fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          Fit: 5 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 2.718282 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -378.2634 </span></span>
<span><span class="co">#&gt; Keeping old fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Done!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fout</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulate_ngs_files/figure-html/unnamed-chunk-7-1.png" width="432"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Estimated proportion misgenotyped</span></span>
<span><span class="va">fout</span><span class="op">$</span><span class="va">prop_mis</span></span>
<span><span class="co">#&gt; [1] 0.07603339</span></span>
<span></span>
<span><span class="co">## Actual proportion misgenotyped</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">fout</span><span class="op">$</span><span class="va">geno</span> <span class="op">!=</span> <span class="va">true_geno</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.07</span></span>
<span></span>
<span><span class="co">## Estimated allele frequency close to true allele frequency</span></span>
<span><span class="va">fout</span><span class="op">$</span><span class="va">par</span><span class="op">$</span><span class="va">alpha</span></span>
<span><span class="co">#&gt; [1] 0.7474485</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Gerard, David, Luís Felipe Ventorim Ferrão, Antonio Augusto Franco
Garcia, and Matthew Stephens. 2018. “Genotyping Polyploids from Messy
Sequencing Data.” <em>Genetics</em> 210 (3). Genetics: 789–807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link uri">https://doi.org/10.1534/genetics.118.301468</a>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
