<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simulate Next-Generation Sequencing Data • updog</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simulate Next-Generation Sequencing Data">
<meta property="og:description" content="updog">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a>
    </li>
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dcgerard/updog/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Simulate Next-Generation Sequencing Data</h1>
                        <h4 data-toc-skip class="author">David Gerard</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/vignettes/simulate_ngs.Rmd" class="external-link"><code>vignettes/simulate_ngs.Rmd</code></a></small>
      <div class="hidden name"><code>simulate_ngs.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<p>We demonstrate how to simulate NGS data under various genotype distributions, then fit these data using <code>flexdog</code>. The genotyping methods are described in Gerard et al. (2018).</p>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p>Let’s suppose that we have 100 hexaploid individuals, with varying levels of read-depth.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dcgerard/updog/" class="external-link">updog</a></span><span class="op">)</span>
<span class="va">nind</span>    <span class="op">&lt;-</span> <span class="fl">100</span>
<span class="va">ploidy</span>  <span class="op">&lt;-</span> <span class="fl">6</span>
<span class="va">sizevec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n   <span class="op">=</span> <span class="va">nind</span>, 
                              min <span class="op">=</span> <span class="fl">50</span>, 
                              max <span class="op">=</span> <span class="fl">200</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We can simulate their read-counts under various genotype distributions, allele biases, overdispersions, and sequencing error rates using the <code>rgeno</code> and <code>rflexdog</code> functions.</p>
<div class="section level3">
<h3 id="f1-population">F1 Population<a class="anchor" aria-label="anchor" href="#f1-population"></a>
</h3>
<p>Suppose these individuals are all siblings where the first parent has 4 copies of the reference allele and the second parent has 5 copies of the reference allele. Then the following code, using <code>rgeno</code>, will simulate the individuals’ genotypes.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">true_geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rgeno.html">rgeno</a></span><span class="op">(</span>n      <span class="op">=</span> <span class="va">nind</span>, 
                   ploidy <span class="op">=</span> <span class="va">ploidy</span>,
                   model  <span class="op">=</span> <span class="st">"f1"</span>, 
                   p1geno <span class="op">=</span> <span class="fl">4</span>,
                   p2geno <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
<p>Once we have their genotypes, we can simulate their read-counts using <code>rflexdog</code>. Let’s suppose that there is a moderate level of allelic bias (0.7) and a small level of overdispersion (0.005). Generally, in the real data that I’ve seen, the bias will range between 0.5 and 2 and the overdispersion will range between 0 and 0.02, with only a few extremely overdispersed SNPs above 0.02.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">refvec</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rflexdog.html">rflexdog</a></span><span class="op">(</span>sizevec <span class="op">=</span> <span class="va">sizevec</span>, 
                   geno    <span class="op">=</span> <span class="va">true_geno</span>, 
                   ploidy  <span class="op">=</span> <span class="va">ploidy</span>, 
                   seq     <span class="op">=</span> <span class="fl">0.001</span>, 
                   bias    <span class="op">=</span> <span class="fl">0.7</span>, 
                   od      <span class="op">=</span> <span class="fl">0.005</span><span class="op">)</span></code></pre></div>
<p>When we plot the data, it looks realistic</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_geno.html">plot_geno</a></span><span class="op">(</span>refvec  <span class="op">=</span> <span class="va">refvec</span>, 
          sizevec <span class="op">=</span> <span class="va">sizevec</span>, 
          ploidy  <span class="op">=</span> <span class="va">ploidy</span>, 
          bias    <span class="op">=</span> <span class="fl">0.7</span>, 
          seq     <span class="op">=</span> <span class="fl">0.001</span>,
          geno    <span class="op">=</span> <span class="va">true_geno</span><span class="op">)</span></code></pre></div>
<p><img src="simulate_ngs_files/figure-html/unnamed-chunk-4-1.png" width="432"></p>
<p>We can test <code>flexdog</code> on these data</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fout</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flexdog.html">flexdog</a></span><span class="op">(</span>refvec  <span class="op">=</span> <span class="va">refvec</span>, 
                sizevec <span class="op">=</span> <span class="va">sizevec</span>, 
                ploidy  <span class="op">=</span> <span class="va">ploidy</span>, 
                model   <span class="op">=</span> <span class="st">"f1"</span><span class="op">)</span>
<span class="co">#&gt;          Fit: 1 of 5 </span>
<span class="co">#&gt; Initial Bias: 0.3678794 </span>
<span class="co">#&gt; Log-Likelihood: -363.9369 </span>
<span class="co">#&gt; Keeping new fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 2 of 5 </span>
<span class="co">#&gt; Initial Bias: 0.6065307 </span>
<span class="co">#&gt; Log-Likelihood: -363.937 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 3 of 5 </span>
<span class="co">#&gt; Initial Bias: 1 </span>
<span class="co">#&gt; Log-Likelihood: -363.9369 </span>
<span class="co">#&gt; Keeping new fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 4 of 5 </span>
<span class="co">#&gt; Initial Bias: 1.648721 </span>
<span class="co">#&gt; Log-Likelihood: -381.6123 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 5 of 5 </span>
<span class="co">#&gt; Initial Bias: 2.718282 </span>
<span class="co">#&gt; Log-Likelihood: -412.8604 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Done!</span></code></pre></div>
<p><code>flexdog</code> gives us reasonable genotyping, and it accurately estimates the proportion of individuals mis-genotyped.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fout</span><span class="op">)</span></code></pre></div>
<p><img src="simulate_ngs_files/figure-html/unnamed-chunk-6-1.png" width="432"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">## Estimated proportion misgenotyped</span>
<span class="va">fout</span><span class="op">$</span><span class="va">prop_mis</span>
<span class="co">#&gt; [1] 0.07011089</span>

<span class="co">## Actual proportion misgenotyped</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">fout</span><span class="op">$</span><span class="va">geno</span> <span class="op">!=</span> <span class="va">true_geno</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.05</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="hwe-population">HWE Population<a class="anchor" aria-label="anchor" href="#hwe-population"></a>
</h3>
<p>Now run the same simulations assuming the individuals are in Hardy-Weinberg population with an allele frequency of 0.75.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">true_geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rgeno.html">rgeno</a></span><span class="op">(</span>n      <span class="op">=</span> <span class="va">nind</span>,
                   ploidy <span class="op">=</span> <span class="va">ploidy</span>,
                   model  <span class="op">=</span> <span class="st">"hw"</span>, 
                   allele_freq <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span>
<span class="va">refvec</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rflexdog.html">rflexdog</a></span><span class="op">(</span>sizevec <span class="op">=</span> <span class="va">sizevec</span>, 
                   geno    <span class="op">=</span> <span class="va">true_geno</span>, 
                   ploidy  <span class="op">=</span> <span class="va">ploidy</span>, 
                   seq     <span class="op">=</span> <span class="fl">0.001</span>, 
                   bias    <span class="op">=</span> <span class="fl">0.7</span>, 
                   od      <span class="op">=</span> <span class="fl">0.005</span><span class="op">)</span>
<span class="va">fout</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flexdog.html">flexdog</a></span><span class="op">(</span>refvec  <span class="op">=</span> <span class="va">refvec</span>, 
                sizevec <span class="op">=</span> <span class="va">sizevec</span>, 
                ploidy  <span class="op">=</span> <span class="va">ploidy</span>, 
                model   <span class="op">=</span> <span class="st">"hw"</span><span class="op">)</span>
<span class="co">#&gt;          Fit: 1 of 5 </span>
<span class="co">#&gt; Initial Bias: 0.3678794 </span>
<span class="co">#&gt; Log-Likelihood: -377.9226 </span>
<span class="co">#&gt; Keeping new fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 2 of 5 </span>
<span class="co">#&gt; Initial Bias: 0.6065307 </span>
<span class="co">#&gt; Log-Likelihood: -377.9226 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 3 of 5 </span>
<span class="co">#&gt; Initial Bias: 1 </span>
<span class="co">#&gt; Log-Likelihood: -377.9226 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 4 of 5 </span>
<span class="co">#&gt; Initial Bias: 1.648721 </span>
<span class="co">#&gt; Log-Likelihood: -377.9226 </span>
<span class="co">#&gt; Keeping new fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 5 of 5 </span>
<span class="co">#&gt; Initial Bias: 2.718282 </span>
<span class="co">#&gt; Log-Likelihood: -377.9226 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Done!</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fout</span><span class="op">)</span></code></pre></div>
<p><img src="simulate_ngs_files/figure-html/unnamed-chunk-7-1.png" width="432"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">## Estimated proportion misgenotyped</span>
<span class="va">fout</span><span class="op">$</span><span class="va">prop_mis</span>
<span class="co">#&gt; [1] 0.07625987</span>

<span class="co">## Actual proportion misgenotyped</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">fout</span><span class="op">$</span><span class="va">geno</span> <span class="op">!=</span> <span class="va">true_geno</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.07</span>

<span class="co">## Estimated allele frequency close to true allele frequency</span>
<span class="va">fout</span><span class="op">$</span><span class="va">par</span><span class="op">$</span><span class="va">alpha</span>
<span class="co">#&gt; [1] 0.7473264</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Gerard, David, Luís Felipe Ventorim Ferrão, Antonio Augusto Franco Garcia, and Matthew Stephens. 2018. “Genotyping Polyploids from Messy Sequencing Data.” <em>Genetics</em> 210 (3). Genetics: 789–807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link uri">https://doi.org/10.1534/genetics.118.301468</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
