<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Example Use of Updog • updog</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Example Use of Updog">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/multidog.html">Genotyping Many SNP's with multidog()</a>
    </li>
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Example Use of Updog</h1>
                        <h4 class="author">David Gerard</h4>
            
      
      
      <div class="hidden name"><code>smells_like_updog.Rmd</code></div>

    </div>

    
    
<div id="whats-updog" class="section level1">
<h1 class="hasAnchor">
<a href="#whats-updog" class="anchor"></a>What’s Updog?</h1>
<p>Updog is a package containing empirical Bayes approaches to genotype individuals (particularly polyploids) from next generation sequencing (NGS) data. We had in mind NGS data that results from a reduced representation library, such as “genotyping-by-sequencing” (GBS) <span class="citation">(Elshire 2011)</span> or “restriction site-associated DNA sequencing” (RAD-seq) <span class="citation">(Baird 2008)</span>.</p>
<p>Updog wields the power of hierarchical modeling to account for some key features of NGS data overlooked in most other analyses, particularly allelic bias and overdispersion. Updog will also automatically account for sequencing errors.</p>
<p>To efficiently account for these features, updog needs to know the distribution of the individual genotypes in the population. The function <code>flexdog</code> can accurately estimate this distribution under a wide variety of situations.</p>
<p>You can read more about the updog method in <span class="citation">Gerard et al. (2018)</span>.</p>
<p>In this vignette, we will go through one example of fitting <code>flexdog</code> on an S1 population of individuals.</p>
</div>
<div id="example-from-an-s1-population" class="section level1">
<h1 class="hasAnchor">
<a href="#example-from-an-s1-population" class="anchor"></a>Example from an S1 Population</h1>
<div id="fit-updog" class="section level2">
<h2 class="hasAnchor">
<a href="#fit-updog" class="anchor"></a>Fit updog</h2>
<p>Load <code>updog</code> and the <code>snpdat</code> dataset. The data frame <code>snpdat</code> contains three example SNPs (single nucleotide polymorphisms) from the study of <span class="citation">Shirasawa et al. (2017)</span>. The individuals in this dataset resulted from a single generation of selfing (an S1 population). You can read more about it by typing <code><a href="../reference/snpdat.html">?snpdat</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">1</span>)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(updog)
<span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"snpdat"</span>)</code></pre></div>
<p>We’ll extract the First SNP.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">smalldat &lt;-<span class="st"> </span>snpdat[snpdat<span class="op">$</span>snp <span class="op">==</span><span class="st"> "SNP1"</span>, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"counts"</span>, <span class="st">"size"</span>, <span class="st">"id"</span>)]
<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(smalldat)
<span class="co">#&gt; # A tibble: 6 x 3</span>
<span class="co">#&gt;   counts  size id           </span>
<span class="co">#&gt;    &lt;int&gt; &lt;int&gt; &lt;chr&gt;        </span>
<span class="co">#&gt; 1    298   354 Xushu18      </span>
<span class="co">#&gt; 2    187   187 Xushu18S1-001</span>
<span class="co">#&gt; 3    201   201 Xushu18S1-002</span>
<span class="co">#&gt; 4    157   184 Xushu18S1-003</span>
<span class="co">#&gt; 5    175   215 Xushu18S1-004</span>
<span class="co">#&gt; 6    283   283 Xushu18S1-005</span></code></pre></div>
<p>We will separate the counts between the children and the parent (the first individual). Note that you do <em>not</em> need the parental counts to fit updog, but they can help improve estimates of the parameters in the updog model.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pref    &lt;-<span class="st"> </span>smalldat<span class="op">$</span>counts[<span class="dv">1</span>]
psize   &lt;-<span class="st"> </span>smalldat<span class="op">$</span>size[<span class="dv">1</span>]
oref    &lt;-<span class="st"> </span>smalldat<span class="op">$</span>counts[<span class="op">-</span><span class="dv">1</span>]
osize   &lt;-<span class="st"> </span>smalldat<span class="op">$</span>size[<span class="op">-</span><span class="dv">1</span>]
ploidy  &lt;-<span class="st"> </span><span class="dv">6</span> <span class="co"># sweet potatoes are hexaploid</span></code></pre></div>
<p>We can first use <code>plot_geno</code> to visualize the raw data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_geno.html">plot_geno</a></span>(<span class="dt">refvec =</span> oref, <span class="dt">sizevec =</span> osize, <span class="dt">ploidy =</span> ploidy)</code></pre></div>
<p><img src="smells_like_updog_files/figure-html/unnamed-chunk-4-1.png" width="432"></p>
<p>Now we use the <code>flexdog</code> function to fit the model. We use <code>model = "s1"</code> because the individuals resulted from one generation of selfing of the same parent.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">uout &lt;-<span class="st"> </span><span class="kw"><a href="../reference/flexdog.html">flexdog</a></span>(<span class="dt">refvec  =</span> oref, 
                <span class="dt">sizevec =</span> osize, 
                <span class="dt">ploidy  =</span> ploidy, 
                <span class="dt">model   =</span> <span class="st">"s1"</span>, 
                <span class="dt">p1ref   =</span> pref, 
                <span class="dt">p1size  =</span> psize)
<span class="co">#&gt;          Fit: 1 of 5 </span>
<span class="co">#&gt; Initial Bias: 0.3678794 </span>
<span class="co">#&gt; Log-Likelihood: -592.9507 </span>
<span class="co">#&gt; Keeping new fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 2 of 5 </span>
<span class="co">#&gt; Initial Bias: 0.6065307 </span>
<span class="co">#&gt; Log-Likelihood: -592.9506 </span>
<span class="co">#&gt; Keeping new fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 3 of 5 </span>
<span class="co">#&gt; Initial Bias: 1 </span>
<span class="co">#&gt; Log-Likelihood: -538.1967 </span>
<span class="co">#&gt; Keeping new fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 4 of 5 </span>
<span class="co">#&gt; Initial Bias: 1.648721 </span>
<span class="co">#&gt; Log-Likelihood: -538.1963 </span>
<span class="co">#&gt; Keeping new fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;          Fit: 5 of 5 </span>
<span class="co">#&gt; Initial Bias: 2.718282 </span>
<span class="co">#&gt; Log-Likelihood: -538.1963 </span>
<span class="co">#&gt; Keeping old fit.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Done!</span></code></pre></div>
</div>
<div id="analyze-output" class="section level2">
<h2 class="hasAnchor">
<a href="#analyze-output" class="anchor"></a>Analyze Output</h2>
<p>We use <code>plot.flexdog</code> to visualize the fit. Points are color coded according to the genotype with the highest posterior probability. For example, a genotype of “4” represents four copies of the reference allele and two copies of the alternative allele (AAAAaa). The level of transparency is proportional to the maximum posterior probability. This is equivalent to the posterior probability that this genotype estimate is correct. The lines represent the mean counts at a given genotype. The “+” symbol with a black dot is the location of the parent.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(uout)</code></pre></div>
<p><img src="smells_like_updog_files/figure-html/unnamed-chunk-6-1.png" width="432"></p>
</div>
<div id="filtering-snps" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-snps" class="anchor"></a>Filtering SNPs</h2>
<p>For downstream analyses, you might want to filter out poorly behaved SNPs. These SNPs might be poorly behaved for a variety of reasons (they might not be real SNPs, it might be much more very difficult to map one allele to the correct location relative to the other allele, etc). Updog gives you some measures to filter out these SNPs.</p>
<p>The most intuitive measure would be the (posterior) proportion of individuals mis-genotyped:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">uout<span class="op">$</span>prop_mis
<span class="co">#&gt; [1] 0.04216397</span></code></pre></div>
<p>For this SNP, we expect about 4.22 percent of the individuals to be mis-genotyped. The specific cutoff you use is context and data dependent. But as a starting point, you could try a loose cutoff by keeping SNPs only if they have a <code>prop_mis</code> of less than 0.2.</p>
<p>From our simulation studies, we also generally get rid of SNPs with overdispersion parameters greater than 0.05 or SNPs with bias parameters either less than 0.5 or greater than 2. However, if you have higher or lower read depths than what we looked at in our simulations, you should adjust these levels accordingly.</p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-baird2008rapid">
<p>Baird, Paul D. AND Atwood, Nathan A. AND Etter. 2008. “Rapid SNP Discovery and Genetic Mapping Using Sequenced RAD Markers.” <em>PLOS ONE</em> 3 (10). Public Library of Science: 1–7. doi:<a href="https://doi.org/10.1371/journal.pone.0003376">10.1371/journal.pone.0003376</a>.</p>
</div>
<div id="ref-elshire2011robust">
<p>Elshire, Jeffrey C. AND Sun, Robert J. AND Glaubitz. 2011. “A Robust, Simple Genotyping-by-Sequencing (GBS) Approach for High Diversity Species.” <em>PLOS ONE</em> 6 (5). Public Library of Science: 1–10. doi:<a href="https://doi.org/10.1371/journal.pone.0019379">10.1371/journal.pone.0019379</a>.</p>
</div>
<div id="ref-gerard2018genotyping">
<p>Gerard, David, Luís Felipe Ventorim Ferrão, Antonio Augusto Franco Garcia, and Matthew Stephens. 2018. “Genotyping Polyploids from Messy Sequencing Data.” <em>Genetics</em> 210 (3). Genetics: 789–807. doi:<a href="https://doi.org/10.1534/genetics.118.301468">10.1534/genetics.118.301468</a>.</p>
</div>
<div id="ref-shirasawa2017high">
<p>Shirasawa, Kenta, Masaru Tanaka, Yasuhiro Takahata, Daifu Ma, Qinghe Cao, Qingchang Liu, Hong Zhai, et al. 2017. “A High-Density SNP Genetic Map Consisting of a Complete Set of Homologous Groups in Autohexaploid Sweetpotato (<em>Ipomoea Batatas</em>).” <em>Scientific Reports</em> 7. Nature Publishing Group. doi:<a href="https://doi.org/10.1038/srep44207">10.1038/srep44207</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#whats-updog">What’s Updog?</a></li>
      <li>
<a href="#example-from-an-s1-population">Example from an S1 Population</a><ul class="nav nav-pills nav-stacked">
<li><a href="#fit-updog">Fit updog</a></li>
      <li><a href="#analyze-output">Analyze Output</a></li>
      <li><a href="#filtering-snps">Filtering SNPs</a></li>
      </ul>
</li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
