<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Example Use of Updog • updog</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Example Use of Updog">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">updog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a></li>
    <li><a class="dropdown-item" href="../articles/oracle_calculations.html">Oracle Calculations</a></li>
    <li><a class="dropdown-item" href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a></li>
    <li><a class="dropdown-item" href="../articles/smells_like_updog.html">Example Use of Updog</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/dcgerard/updog/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Example Use of Updog</h1>
                        <h4 data-toc-skip class="author">David
Gerard</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/vignettes/smells_like_updog.Rmd" class="external-link"><code>vignettes/smells_like_updog.Rmd</code></a></small>
      <div class="d-none name"><code>smells_like_updog.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="whats-updog">What’s Updog?<a class="anchor" aria-label="anchor" href="#whats-updog"></a>
</h2>
<p>Updog is a package containing empirical Bayes approaches to genotype
individuals (particularly polyploids) from next generation sequencing
(NGS) data. We had in mind NGS data that results from a reduced
representation library, such as “genotyping-by-sequencing” (GBS)
(Elshire et al., 2011) or “restriction site-associated DNA sequencing”
(RAD-seq) (Baird et al., 2008).</p>
<p>Updog wields the power of hierarchical modeling to account for some
key features of NGS data overlooked in most other analyses, particularly
allelic bias and overdispersion. Updog will also automatically account
for sequencing errors.</p>
<p>To efficiently account for these features, updog needs to know the
distribution of the individual genotypes in the population. The function
<code>flexdog</code> can accurately estimate this distribution under a
wide variety of situations.</p>
<p>You can read more about the updog method in Gerard et al. (2018).</p>
<p>In this vignette, we will go through one example of fitting
<code>flexdog</code> on an S1 population of individuals.</p>
</div>
<div class="section level2">
<h2 id="example-from-an-s1-population">Example from an S1 Population<a class="anchor" aria-label="anchor" href="#example-from-an-s1-population"></a>
</h2>
<div class="section level3">
<h3 id="fit-updog">Fit updog<a class="anchor" aria-label="anchor" href="#fit-updog"></a>
</h3>
<p>Load <code>updog</code> and the <code>snpdat</code> dataset. The data
frame <code>snpdat</code> contains three example SNPs (single nucleotide
polymorphisms) from the study of Shirasawa et al. (2017). The
individuals in this dataset resulted from a single generation of selfing
(an S1 population). You can read more about it by typing
<code><a href="../reference/snpdat.html">?snpdat</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dcgerard.github.io/updog/">updog</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"snpdat"</span><span class="op">)</span></span></code></pre></div>
<p>We’ll extract the First SNP.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">smalldat</span> <span class="op">&lt;-</span> <span class="va">snpdat</span><span class="op">[</span><span class="va">snpdat</span><span class="op">$</span><span class="va">snp</span> <span class="op">==</span> <span class="st">"SNP1"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"counts"</span>, <span class="st">"size"</span>, <span class="st">"id"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">smalldat</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt;   counts  size id           </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>    298   354 Xushu18      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>    187   187 Xushu18S1-001</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>    201   201 Xushu18S1-002</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>    157   184 Xushu18S1-003</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>    175   215 Xushu18S1-004</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>    283   283 Xushu18S1-005</span></span></code></pre></div>
<p>We will separate the counts between the children and the parent (the
first individual). Note that you do <em>not</em> need the parental
counts to fit updog, but they can help improve estimates of the
parameters in the updog model.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pref</span>    <span class="op">&lt;-</span> <span class="va">smalldat</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">psize</span>   <span class="op">&lt;-</span> <span class="va">smalldat</span><span class="op">$</span><span class="va">size</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">oref</span>    <span class="op">&lt;-</span> <span class="va">smalldat</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">osize</span>   <span class="op">&lt;-</span> <span class="va">smalldat</span><span class="op">$</span><span class="va">size</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">ploidy</span>  <span class="op">&lt;-</span> <span class="fl">6</span> <span class="co"># sweet potatoes are hexaploid</span></span></code></pre></div>
<p>We can first use <code>plot_geno</code> to visualize the raw
data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_geno.html">plot_geno</a></span><span class="op">(</span>refvec <span class="op">=</span> <span class="va">oref</span>, sizevec <span class="op">=</span> <span class="va">osize</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span><span class="op">)</span></span></code></pre></div>
<p><img src="smells_like_updog_files/figure-html/unnamed-chunk-4-1.png" class="r-plt" alt="" width="432"></p>
<p>Now we use the <code>flexdog</code> function to fit the model. We use
<code>model = "s1"</code> because the individuals resulted from one
generation of selfing of the same parent.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">uout</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flexdog.html">flexdog</a></span><span class="op">(</span>refvec  <span class="op">=</span> <span class="va">oref</span>, </span>
<span>                sizevec <span class="op">=</span> <span class="va">osize</span>, </span>
<span>                ploidy  <span class="op">=</span> <span class="va">ploidy</span>, </span>
<span>                model   <span class="op">=</span> <span class="st">"s1"</span>, </span>
<span>                p1ref   <span class="op">=</span> <span class="va">pref</span>, </span>
<span>                p1size  <span class="op">=</span> <span class="va">psize</span><span class="op">)</span></span>
<span><span class="co">#&gt;          Fit: 1 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 0.3678794 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -593.6623 </span></span>
<span><span class="co">#&gt; Keeping new fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          Fit: 2 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 0.6065307 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -593.6623 </span></span>
<span><span class="co">#&gt; Keeping old fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          Fit: 3 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 1 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -538.1835 </span></span>
<span><span class="co">#&gt; Keeping new fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          Fit: 4 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 1.648721 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -538.1831 </span></span>
<span><span class="co">#&gt; Keeping new fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          Fit: 5 of 5 </span></span>
<span><span class="co">#&gt; Initial Bias: 2.718282 </span></span>
<span><span class="co">#&gt; Log-Likelihood: -538.1831 </span></span>
<span><span class="co">#&gt; Keeping old fit.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Done!</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="analyze-output">Analyze Output<a class="anchor" aria-label="anchor" href="#analyze-output"></a>
</h3>
<p>We use <code>plot.flexdog</code> to visualize the fit. Points are
color coded according to the genotype with the highest posterior
probability. For example, a genotype of “4” represents four copies of
the reference allele and two copies of the alternative allele (AAAAaa).
The level of transparency is proportional to the maximum posterior
probability. This is equivalent to the posterior probability that this
genotype estimate is correct. The lines represent the mean counts at a
given genotype. The “+” symbol with a black dot is the location of the
parent.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">uout</span><span class="op">)</span></span></code></pre></div>
<p><img src="smells_like_updog_files/figure-html/unnamed-chunk-6-1.png" class="r-plt" alt="" width="432"></p>
</div>
<div class="section level3">
<h3 id="filtering-snps">Filtering SNPs<a class="anchor" aria-label="anchor" href="#filtering-snps"></a>
</h3>
<p>For downstream analyses, you might want to filter out poorly behaved
SNPs. These SNPs might be poorly behaved for a variety of reasons (they
might not be real SNPs, it might be much more very difficult to map one
allele to the correct location relative to the other allele, etc). Updog
gives you some measures to filter out these SNPs.</p>
<p>The most intuitive measure would be the (posterior) proportion of
individuals mis-genotyped:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">uout</span><span class="op">$</span><span class="va">prop_mis</span></span>
<span><span class="co">#&gt; [1] 0.04220289</span></span></code></pre></div>
<p>For this SNP, we expect about 4.22 percent of the individuals to be
mis-genotyped. The specific cutoff you use is context and data
dependent. But as a starting point, you could try a loose cutoff by
keeping SNPs only if they have a <code>prop_mis</code> of less than
0.2.</p>
<p>From our simulation studies, we also generally get rid of SNPs with
overdispersion parameters greater than 0.05 or SNPs with bias parameters
either less than 0.5 or greater than 2. However, if you have higher or
lower read depths than what we looked at in our simulations, you should
adjust these levels accordingly.</p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Baird, Paul D. AND Atwood, Nathan A. AND Etter. 2008. “Rapid SNP
Discovery and Genetic Mapping Using Sequenced RAD Markers.” <em>PLOS
ONE</em> 3 (10). Public Library of Science: 1–7. <a href="https://doi.org/10.1371/journal.pone.0003376" class="external-link uri">https://doi.org/10.1371/journal.pone.0003376</a>.</p>
<p>Elshire, Jeffrey C. AND Sun, Robert J. AND Glaubitz. 2011. “A Robust,
Simple Genotyping-by-Sequencing (GBS) Approach for High Diversity
Species.” <em>PLOS ONE</em> 6 (5). Public Library of Science: 1–10. <a href="https://doi.org/10.1371/journal.pone.0019379" class="external-link uri">https://doi.org/10.1371/journal.pone.0019379</a>.</p>
<p>Gerard, David, Luís Felipe Ventorim Ferrão, Antonio Augusto Franco
Garcia, and Matthew Stephens. 2018. “Genotyping Polyploids from Messy
Sequencing Data.” <em>Genetics</em> 210 (3). Genetics: 789–807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link uri">https://doi.org/10.1534/genetics.118.301468</a>.</p>
<p>Shirasawa, Kenta, Masaru Tanaka, Yasuhiro Takahata, Daifu Ma, Qinghe
Cao, Qingchang Liu, Hong Zhai, et al. 2017. “A High-Density SNP Genetic
Map Consisting of a Complete Set of Homologous Groups in Autohexaploid
Sweetpotato (<em>Ipomoea batatas</em>).” <em>Scientific Reports</em> 7.
Nature Publishing Group. <a href="https://doi.org/10.1038/srep44207" class="external-link uri">https://doi.org/10.1038/srep44207</a>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
