<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Returns the oracle misclassification rates for each genotype. — oracle_mis_vec • updog</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Returns the oracle misclassification rates for each genotype. — oracle_mis_vec"><meta property="og:description" content="Given perfect knowledge of the data generating parameters,
oracle_mis_vec calculates the misclassification error
rate at each genotype. This differs from oracle_mis
in that this will not average over the genotype distribution to
get an overall misclassification error rate. That is, oracle_mis_vec
returns a vector of misclassification error rates conditional on
each genotype."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a>
    </li>
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/updog/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Returns the oracle misclassification rates for each genotype.</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/R/RcppExports.R" class="external-link"><code>R/RcppExports.R</code></a></small>
    <div class="hidden name"><code>oracle_mis_vec.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given perfect knowledge of the data generating parameters,
<code>oracle_mis_vec</code> calculates the misclassification error
rate at each genotype. This differs from <code><a href="oracle_mis.html">oracle_mis</a></code>
in that this will <em>not</em> average over the genotype distribution to
get an overall misclassification error rate. That is, <code>oracle_mis_vec</code>
returns a vector of misclassification error rates <em>conditional</em> on
each genotype.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">oracle_mis_vec</span><span class="op">(</span><span class="va">n</span>, <span class="va">ploidy</span>, <span class="va">seq</span>, <span class="va">bias</span>, <span class="va">od</span>, <span class="va">dist</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>n</dt>
<dd><p>The read-depth.</p></dd>


<dt>ploidy</dt>
<dd><p>The ploidy of the individual.</p></dd>


<dt>seq</dt>
<dd><p>The sequencing error rate.</p></dd>


<dt>bias</dt>
<dd><p>The allele-bias.</p></dd>


<dt>od</dt>
<dd><p>The overdispersion parameter.</p></dd>


<dt>dist</dt>
<dd><p>The distribution of the alleles.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A vector of numerics. Element i is the oracle misclassification
    error rate when genotyping an individual with actual
    genotype i + 1.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This is an ideal level of the misclassification error rate and
any real method will have a larger rate than this. This is a useful
approximation when you have a lot of individuals.</p>
<p>To come up with <code>dist</code>, you need some additional assumptions.
For example, if the population is in Hardy-Weinberg equilibrium and
the allele frequency is <code>alpha</code> then you could calculate
<code>dist</code> using the R code: <code>dbinom(x = 0:ploidy, size = ploidy, prob = alpha)</code>.
Alternatively, if you know the genotypes of the individual's two parents are, say,
<code>ref_count1</code> and <code>ref_count2</code>, then you could use the <code><a href="get_q_array.html">get_q_array</a></code>
function from the updog package: <code>get_q_array(ploidy)[ref_count1 + 1, ref_count2 + 1, ]</code>.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    
<ul><li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">doi:10.1534/genetics.118.301468</a>
.</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Hardy-Weinberg population with allele-frequency of 0.75.</span></span></span>
<span class="r-in"><span><span class="co">## Moderate bias and moderate overdispersion.</span></span></span>
<span class="r-in"><span><span class="va">ploidy</span> <span class="op">&lt;-</span> <span class="fl">4</span></span></span>
<span class="r-in"><span><span class="va">dist</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">dbinom</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="va">ploidy</span>, <span class="va">ploidy</span>, <span class="fl">0.75</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">om</span> <span class="op">&lt;-</span> <span class="fu">oracle_mis_vec</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.001</span>,</span></span>
<span class="r-in"><span>                     bias <span class="op">=</span> <span class="fl">0.7</span>, od <span class="op">=</span> <span class="fl">0.01</span>, dist <span class="op">=</span> <span class="va">dist</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">om</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0001497595 0.0657395175 0.0702925658 0.0267344300 0.0008221220</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get same output as oracle_mis this way:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">dist</span> <span class="op">*</span> <span class="va">om</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02944818</span>
<span class="r-in"><span><span class="fu"><a href="oracle_mis.html">oracle_mis</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.001</span>,</span></span>
<span class="r-in"><span>           bias <span class="op">=</span> <span class="fl">0.7</span>, od <span class="op">=</span> <span class="fl">0.01</span>, dist <span class="op">=</span> <span class="va">dist</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02944818</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

