<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Returns the oracle misclassification rates for each genotype. — oracle_mis_vec • updog</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Returns the oracle misclassification rates for each genotype. — oracle_mis_vec"><meta name="description" content="Given perfect knowledge of the data generating parameters,
oracle_mis_vec calculates the misclassification error
rate at each genotype. This differs from oracle_mis
in that this will not average over the genotype distribution to
get an overall misclassification error rate. That is, oracle_mis_vec
returns a vector of misclassification error rates conditional on
each genotype."><meta property="og:description" content="Given perfect knowledge of the data generating parameters,
oracle_mis_vec calculates the misclassification error
rate at each genotype. This differs from oracle_mis
in that this will not average over the genotype distribution to
get an overall misclassification error rate. That is, oracle_mis_vec
returns a vector of misclassification error rates conditional on
each genotype."><meta property="og:image" content="/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">updog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a></li>
    <li><a class="dropdown-item" href="../articles/oracle_calculations.html">Oracle Calculations</a></li>
    <li><a class="dropdown-item" href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a></li>
    <li><a class="dropdown-item" href="../articles/smells_like_updog.html">Example Use of Updog</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/dcgerard/updog/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Returns the oracle misclassification rates for each genotype.</h1>
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/R/RcppExports.R" class="external-link"><code>R/RcppExports.R</code></a></small>
      <div class="d-none name"><code>oracle_mis_vec.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Given perfect knowledge of the data generating parameters,
<code>oracle_mis_vec</code> calculates the misclassification error
rate at each genotype. This differs from <code><a href="oracle_mis.html">oracle_mis</a></code>
in that this will <em>not</em> average over the genotype distribution to
get an overall misclassification error rate. That is, <code>oracle_mis_vec</code>
returns a vector of misclassification error rates <em>conditional</em> on
each genotype.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">oracle_mis_vec</span><span class="op">(</span><span class="va">n</span>, <span class="va">ploidy</span>, <span class="va">seq</span>, <span class="va">bias</span>, <span class="va">od</span>, <span class="va">dist</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-n">n<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>The read-depth.</p></dd>


<dt id="arg-ploidy">ploidy<a class="anchor" aria-label="anchor" href="#arg-ploidy"></a></dt>
<dd><p>The ploidy of the individual.</p></dd>


<dt id="arg-seq">seq<a class="anchor" aria-label="anchor" href="#arg-seq"></a></dt>
<dd><p>The sequencing error rate.</p></dd>


<dt id="arg-bias">bias<a class="anchor" aria-label="anchor" href="#arg-bias"></a></dt>
<dd><p>The allele-bias.</p></dd>


<dt id="arg-od">od<a class="anchor" aria-label="anchor" href="#arg-od"></a></dt>
<dd><p>The overdispersion parameter.</p></dd>


<dt id="arg-dist">dist<a class="anchor" aria-label="anchor" href="#arg-dist"></a></dt>
<dd><p>The distribution of the alleles.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A vector of numerics. Element i is the oracle misclassification
    error rate when genotyping an individual with actual
    genotype i + 1.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This is an ideal level of the misclassification error rate and
any real method will have a larger rate than this. This is a useful
approximation when you have a lot of individuals.</p>
<p>To come up with <code>dist</code>, you need some additional assumptions.
For example, if the population is in Hardy-Weinberg equilibrium and
the allele frequency is <code>alpha</code> then you could calculate
<code>dist</code> using the R code: <code>dbinom(x = 0:ploidy, size = ploidy, prob = alpha)</code>.
Alternatively, if you know the genotypes of the individual's two parents are, say,
<code>ref_count1</code> and <code>ref_count2</code>, then you could use the <code><a href="get_q_array.html">get_q_array</a></code>
function from the updog package: <code>get_q_array(ploidy)[ref_count1 + 1, ref_count2 + 1, ]</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>

<ul><li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">doi:10.1534/genetics.118.301468</a>
.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>David Gerard</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Hardy-Weinberg population with allele-frequency of 0.75.</span></span></span>
<span class="r-in"><span><span class="co">## Moderate bias and moderate overdispersion.</span></span></span>
<span class="r-in"><span><span class="va">ploidy</span> <span class="op">&lt;-</span> <span class="fl">4</span></span></span>
<span class="r-in"><span><span class="va">dist</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">dbinom</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="va">ploidy</span>, <span class="va">ploidy</span>, <span class="fl">0.75</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">om</span> <span class="op">&lt;-</span> <span class="fu">oracle_mis_vec</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.001</span>,</span></span>
<span class="r-in"><span>                     bias <span class="op">=</span> <span class="fl">0.7</span>, od <span class="op">=</span> <span class="fl">0.01</span>, dist <span class="op">=</span> <span class="va">dist</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">om</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0001497595 0.0657395175 0.0702925658 0.0267344300 0.0008221220</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get same output as oracle_mis this way:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">dist</span> <span class="op">*</span> <span class="va">om</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02944818</span>
<span class="r-in"><span><span class="fu"><a href="oracle_mis.html">oracle_mis</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.001</span>,</span></span>
<span class="r-in"><span>           bias <span class="op">=</span> <span class="fl">0.7</span>, od <span class="op">=</span> <span class="fl">0.01</span>, dist <span class="op">=</span> <span class="va">dist</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02944818</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

