<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This returns the joint distribution of the true genotypes and an oracle
estimator given perfect knowledge of the data generating process. This is a useful
approximation when you have a lot of individuals."><title>The joint probability of the genotype and the genotype estimate
of an oracle estimator. — oracle_joint • updog</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="The joint probability of the genotype and the genotype estimate
of an oracle estimator. — oracle_joint"><meta property="og:description" content="This returns the joint distribution of the true genotypes and an oracle
estimator given perfect knowledge of the data generating process. This is a useful
approximation when you have a lot of individuals."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">updog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a>
    <a class="dropdown-item" href="../articles/oracle_calculations.html">Oracle Calculations</a>
    <a class="dropdown-item" href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    <a class="dropdown-item" href="../articles/smells_like_updog.html">Example Use of Updog</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/dcgerard/updog/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>The joint probability of the genotype and the genotype estimate
of an oracle estimator.</h1>
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/R/RcppExports.R" class="external-link"><code>R/RcppExports.R</code></a></small>
      <div class="d-none name"><code>oracle_joint.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This returns the joint distribution of the true genotypes and an oracle
estimator given perfect knowledge of the data generating process. This is a useful
approximation when you have a lot of individuals.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">oracle_joint</span><span class="op">(</span><span class="va">n</span>, <span class="va">ploidy</span>, <span class="va">seq</span>, <span class="va">bias</span>, <span class="va">od</span>, <span class="va">dist</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>n</dt>
<dd><p>The read-depth.</p></dd>


<dt>ploidy</dt>
<dd><p>The ploidy of the individual.</p></dd>


<dt>seq</dt>
<dd><p>The sequencing error rate.</p></dd>


<dt>bias</dt>
<dd><p>The allele-bias.</p></dd>


<dt>od</dt>
<dd><p>The overdispersion parameter.</p></dd>


<dt>dist</dt>
<dd><p>The distribution of the alleles.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A matrix. Element (i, j) is the joint probability of estimating
    the genotype to be i+1 when the true genotype is j+1. That is, the
    estimated genotype indexes the rows and the true genotype indexes
    the columns. This is when
    using an oracle estimator.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>To come up with <code>dist</code>, you need some additional assumptions.
For example, if the population is in Hardy-Weinberg equilibrium and
the allele frequency is <code>alpha</code> then you could calculate
<code>dist</code> using the R code: <code>dbinom(x = 0:ploidy, size = ploidy, prob = alpha)</code>.
Alternatively, if you know the genotypes of the individual's two parents are, say,
<code>ref_count1</code> and <code>ref_count2</code>, then you could use the <code><a href="get_q_array.html">get_q_array</a></code>
function from the updog package: <code>get_q_array(ploidy)[ref_count1 + 1, ref_count2 + 1, ]</code>.</p>
<p>See the Examples to see how to reconcile the output of <code>oracle_joint</code>
with <code><a href="oracle_mis.html">oracle_mis</a></code> and <code><a href="oracle_mis_vec.html">oracle_mis_vec</a></code>.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    
<ul><li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">doi:10.1534/genetics.118.301468</a>
.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index">
<dl><dt><code><a href="oracle_plot.html">oracle_plot</a></code></dt>
<dd><p>For visualizing the joint distribution output from <code>oracle_joint</code>.</p></dd>

  <dt><code><a href="oracle_mis_from_joint.html">oracle_mis_from_joint</a></code></dt>
<dd><p>For obtaining the same results as <code><a href="oracle_mis.html">oracle_mis</a></code>
      directly from the output of <code>oracle_joint</code>.</p></dd>

  <dt><code><a href="oracle_mis_vec_from_joint.html">oracle_mis_vec_from_joint</a></code></dt>
<dd><p>For obtaining the same results as <code><a href="oracle_mis_vec.html">oracle_mis_vec</a></code>
      directly from the output of <code>oracle_joint</code>.</p></dd>

  <dt><code><a href="oracle_cor_from_joint.html">oracle_cor_from_joint</a></code></dt>
<dd><p>For obtaining the same results as <code><a href="oracle_cor.html">oracle_cor</a></code>
      directly from the output of <code>oracle_joint</code>.</p></dd>


</dl></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>David Gerard</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Hardy-Weinberg population with allele-frequency of 0.75.</span></span></span>
<span class="r-in"><span><span class="co">## Moderate bias and moderate overdispersion.</span></span></span>
<span class="r-in"><span><span class="va">ploidy</span> <span class="op">&lt;-</span> <span class="fl">4</span></span></span>
<span class="r-in"><span><span class="va">dist</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">dbinom</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="va">ploidy</span>, <span class="va">ploidy</span>, <span class="fl">0.75</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">jd</span> <span class="op">&lt;-</span> <span class="fu">oracle_joint</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.001</span>,</span></span>
<span class="r-in"><span>                   bias <span class="op">=</span> <span class="fl">0.7</span>, od <span class="op">=</span> <span class="fl">0.01</span>, dist <span class="op">=</span> <span class="va">dist</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">jd</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              [,1]         [,2]         [,3]         [,4]         [,5]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 3.905665e-03 1.759022e-07 8.379767e-17 1.784566e-29 3.601827e-52</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 5.849980e-07 4.379346e-02 2.180335e-03 2.159655e-09 3.235599e-26</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,] 1.897235e-20 3.081362e-03 1.961102e-01 1.099225e-02 6.173803e-14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] 1.314974e-34 2.427440e-09 1.264700e-02 4.105964e-01 2.601245e-04</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,] 2.284980e-57 7.543647e-25 9.090651e-12 2.863373e-04 3.161461e-01</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get same output as oracle_mis this way:</span></span></span>
<span class="r-in"><span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">jd</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02944818</span>
<span class="r-in"><span><span class="fu"><a href="oracle_mis.html">oracle_mis</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.001</span>,</span></span>
<span class="r-in"><span>           bias <span class="op">=</span> <span class="fl">0.7</span>, od <span class="op">=</span> <span class="fl">0.01</span>, dist <span class="op">=</span> <span class="va">dist</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02944818</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get same output as oracle_mis_vec this way:</span></span></span>
<span class="r-in"><span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">jd</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, STATS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">jd</span><span class="op">)</span>, FUN <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0001497595 0.0657395175 0.0702925658 0.0267344300 0.0008221220</span>
<span class="r-in"><span><span class="fu"><a href="oracle_mis_vec.html">oracle_mis_vec</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.001</span>,</span></span>
<span class="r-in"><span>               bias <span class="op">=</span> <span class="fl">0.7</span>, od <span class="op">=</span> <span class="fl">0.01</span>, dist <span class="op">=</span> <span class="va">dist</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0001497595 0.0657395175 0.0702925658 0.0267344300 0.0008221220</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

