<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The x-axis is the counts of the non-reference allele,
and the y-axis is the counts of the reference allele.
Transparency is controlled by the maxpostprob
vector. These types of plots are used in Gerard et. al. (2018) and
Gerard and Ferrão (2020)."><title>Make a genotype plot. — plot_geno • updog</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Make a genotype plot. — plot_geno"><meta property="og:description" content="The x-axis is the counts of the non-reference allele,
and the y-axis is the counts of the reference allele.
Transparency is controlled by the maxpostprob
vector. These types of plots are used in Gerard et. al. (2018) and
Gerard and Ferrão (2020)."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">updog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a>
    <a class="dropdown-item" href="../articles/oracle_calculations.html">Oracle Calculations</a>
    <a class="dropdown-item" href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    <a class="dropdown-item" href="../articles/smells_like_updog.html">Example Use of Updog</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/dcgerard/updog/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Make a genotype plot.</h1>
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/R/plot_geno.R" class="external-link"><code>R/plot_geno.R</code></a></small>
      <div class="d-none name"><code>plot_geno.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The x-axis is the counts of the non-reference allele,
and the y-axis is the counts of the reference allele.
Transparency is controlled by the <code>maxpostprob</code>
vector. These types of plots are used in Gerard et. al. (2018) and
Gerard and Ferrão (2020).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_geno</span><span class="op">(</span></span>
<span>  <span class="va">refvec</span>,</span>
<span>  <span class="va">sizevec</span>,</span>
<span>  <span class="va">ploidy</span>,</span>
<span>  p1ref <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p1size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p2ref <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p2size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  geno <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seq <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  bias <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  maxpostprob <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p1geno <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p2geno <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_colorblind <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>refvec</dt>
<dd><p>A vector of non-negative integers. The number of
reference reads observed in the individuals</p></dd>


<dt>sizevec</dt>
<dd><p>A vector of positive integers. The total number of
reads in the individuals.</p></dd>


<dt>ploidy</dt>
<dd><p>A non-negative integer. The ploidy of the species.</p></dd>


<dt>p1ref</dt>
<dd><p>A vector of non-negative integers. The number of
reference reads observed in parent 1 (if the individuals are
all siblings).</p></dd>


<dt>p1size</dt>
<dd><p>A vector of positive integers. The total number of
reads in parent 1 (if the individuals are
all siblings).</p></dd>


<dt>p2ref</dt>
<dd><p>A vector of non-negative integers. The number of
reference reads observed in parent 2 (if the individuals are
all siblings).</p></dd>


<dt>p2size</dt>
<dd><p>A vector of positive integers. The total number of
reads in parent 2 (if the individuals are
all siblings).</p></dd>


<dt>geno</dt>
<dd><p>The individual genotypes.</p></dd>


<dt>seq</dt>
<dd><p>The sequencing error rate.</p></dd>


<dt>bias</dt>
<dd><p>The bias parameter.</p></dd>


<dt>maxpostprob</dt>
<dd><p>A vector of the posterior probabilities of being at
the modal genotype.</p></dd>


<dt>p1geno</dt>
<dd><p>Parent 1's genotype.</p></dd>


<dt>p2geno</dt>
<dd><p>Parent 2's genotype.</p></dd>


<dt>use_colorblind</dt>
<dd><p>A logical. Should we use a colorblind safe palette (<code>TRUE</code>),
or not (<code>FALSE</code>)? Only allowed if <code>ploidy &lt;= 6</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></code> object for the genotype plot.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>If parental genotypes are provided (<code>p1geno</code> and <code>p2geno</code>) then
they will be colored the same as the offspring. Since they are often hard to see,
a small black dot will also indicate their position.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    
<ul><li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">doi:10.1534/genetics.118.301468</a>
.</p></li>
<li><p>Gerard, David, and Luís Felipe Ventorim Ferrão. "Priors for genotyping polyploids." Bioinformatics 36, no. 6 (2020): 1795-1800. <a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link">doi:10.1093/bioinformatics/btz852</a>
.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>David Gerard</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"snpdat"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">refvec</span>  <span class="op">&lt;-</span> <span class="va">snpdat</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="va">snpdat</span><span class="op">$</span><span class="va">snp</span> <span class="op">==</span> <span class="st">"SNP1"</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">sizevec</span> <span class="op">&lt;-</span> <span class="va">snpdat</span><span class="op">$</span><span class="va">size</span><span class="op">[</span><span class="va">snpdat</span><span class="op">$</span><span class="va">snp</span> <span class="op">==</span> <span class="st">"SNP1"</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">ploidy</span>  <span class="op">&lt;-</span> <span class="fl">6</span></span></span>
<span class="r-in"><span><span class="fu">plot_geno</span><span class="op">(</span>refvec <span class="op">=</span> <span class="va">refvec</span>, sizevec <span class="op">=</span> <span class="va">sizevec</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot_geno-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

