<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Make a genotype plot. — plot_geno • updog</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Make a genotype plot. — plot_geno"><meta property="og:description" content="The x-axis is the counts of the non-reference allele,
and the y-axis is the counts of the reference allele.
Transparency is controlled by the maxpostprob
vector. These types of plots are used in Gerard et. al. (2018) and
Gerard and Ferrão (2020)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a>
    </li>
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/updog/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make a genotype plot.</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/R/plot_geno.R" class="external-link"><code>R/plot_geno.R</code></a></small>
    <div class="hidden name"><code>plot_geno.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The x-axis is the counts of the non-reference allele,
and the y-axis is the counts of the reference allele.
Transparency is controlled by the <code>maxpostprob</code>
vector. These types of plots are used in Gerard et. al. (2018) and
Gerard and Ferrão (2020).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">plot_geno</span><span class="op">(</span>
  <span class="va">refvec</span>,
  <span class="va">sizevec</span>,
  <span class="va">ploidy</span>,
  p1ref <span class="op">=</span> <span class="cn">NULL</span>,
  p1size <span class="op">=</span> <span class="cn">NULL</span>,
  p2ref <span class="op">=</span> <span class="cn">NULL</span>,
  p2size <span class="op">=</span> <span class="cn">NULL</span>,
  geno <span class="op">=</span> <span class="cn">NULL</span>,
  seq <span class="op">=</span> <span class="fl">0</span>,
  bias <span class="op">=</span> <span class="fl">1</span>,
  maxpostprob <span class="op">=</span> <span class="cn">NULL</span>,
  p1geno <span class="op">=</span> <span class="cn">NULL</span>,
  p2geno <span class="op">=</span> <span class="cn">NULL</span>,
  use_colorblind <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>refvec</dt>
<dd><p>A vector of non-negative integers. The number of
reference reads observed in the individuals</p></dd>
<dt>sizevec</dt>
<dd><p>A vector of positive integers. The total number of
reads in the individuals.</p></dd>
<dt>ploidy</dt>
<dd><p>A non-negative integer. The ploidy of the species.</p></dd>
<dt>p1ref</dt>
<dd><p>A vector of non-negative integers. The number of
reference reads observed in parent 1 (if the individuals are
all siblings).</p></dd>
<dt>p1size</dt>
<dd><p>A vector of positive integers. The total number of
reads in parent 1 (if the individuals are
all siblings).</p></dd>
<dt>p2ref</dt>
<dd><p>A vector of non-negative integers. The number of
reference reads observed in parent 2 (if the individuals are
all siblings).</p></dd>
<dt>p2size</dt>
<dd><p>A vector of positive integers. The total number of
reads in parent 2 (if the individuals are
all siblings).</p></dd>
<dt>geno</dt>
<dd><p>The individual genotypes.</p></dd>
<dt>seq</dt>
<dd><p>The sequencing error rate.</p></dd>
<dt>bias</dt>
<dd><p>The bias parameter.</p></dd>
<dt>maxpostprob</dt>
<dd><p>A vector of the posterior probabilities of being at
the modal genotype.</p></dd>
<dt>p1geno</dt>
<dd><p>Parent 1's genotype.</p></dd>
<dt>p2geno</dt>
<dd><p>Parent 2's genotype.</p></dd>
<dt>use_colorblind</dt>
<dd><p>A logical. Should we use a colorblind safe palette (<code>TRUE</code>),
or not (<code>FALSE</code>)? Only allowed if <code>ploidy &lt;= 6</code>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></code> object for the genotype plot.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>If parental genotypes are provided (<code>p1geno</code> and <code>p2geno</code>) then
they will be colored the same as the offspring. Since they are often hard to see,
a small black dot will also indicate their position.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    
<ul><li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. doi: <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">10.1534/genetics.118.301468</a>
.</p></li>
<li><p>Gerard, David, and Luís Felipe Ventorim Ferrão. "Priors for genotyping polyploids." Bioinformatics 36, no. 6 (2020): 1795-1800. doi: <a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link">10.1093/bioinformatics/btz852</a>
.</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"snpdat"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">refvec</span>  <span class="op">&lt;-</span> <span class="va">snpdat</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="va">snpdat</span><span class="op">$</span><span class="va">snp</span> <span class="op">==</span> <span class="st">"SNP1"</span><span class="op">]</span></span>
<span class="r-in"><span class="va">sizevec</span> <span class="op">&lt;-</span> <span class="va">snpdat</span><span class="op">$</span><span class="va">size</span><span class="op">[</span><span class="va">snpdat</span><span class="op">$</span><span class="va">snp</span> <span class="op">==</span> <span class="st">"SNP1"</span><span class="op">]</span></span>
<span class="r-in"><span class="va">ploidy</span>  <span class="op">&lt;-</span> <span class="fl">6</span></span>
<span class="r-in"><span class="fu">plot_geno</span><span class="op">(</span>refvec <span class="op">=</span> <span class="va">refvec</span>, sizevec <span class="op">=</span> <span class="va">sizevec</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span><span class="op">)</span></span>
<span class="r-plt img"><img src="plot_geno-1.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

