<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Simulate GBS data from the flexdog likelihood. — rflexdog • updog</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate GBS data from the flexdog likelihood. — rflexdog"><meta name="description" content="This will take a vector of genotypes and a vector of total read-counts, then generate a vector of reference
counts. To get the genotypes, you could use rgeno. The likelihood used to generate read-counts
is described in detail in Gerard et. al. (2018)."><meta property="og:description" content="This will take a vector of genotypes and a vector of total read-counts, then generate a vector of reference
counts. To get the genotypes, you could use rgeno. The likelihood used to generate read-counts
is described in detail in Gerard et. al. (2018)."><meta property="og:image" content="https://dcgerard.github.io/updog/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">updog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a></li>
    <li><a class="dropdown-item" href="../articles/oracle_calculations.html">Oracle Calculations</a></li>
    <li><a class="dropdown-item" href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a></li>
    <li><a class="dropdown-item" href="../articles/smells_like_updog.html">Example Use of Updog</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/dcgerard/updog/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Simulate GBS data from the <code><a href="flexdog.html">flexdog</a></code> likelihood.</h1>
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/R/rflexdog.R" class="external-link"><code>R/rflexdog.R</code></a></small>
      <div class="d-none name"><code>rflexdog.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This will take a vector of genotypes and a vector of total read-counts, then generate a vector of reference
counts. To get the genotypes, you could use <code><a href="rgeno.html">rgeno</a></code>. The likelihood used to generate read-counts
is described in detail in Gerard et. al. (2018).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rflexdog</span><span class="op">(</span><span class="va">sizevec</span>, <span class="va">geno</span>, <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.005</span>, bias <span class="op">=</span> <span class="fl">1</span>, od <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sizevec">sizevec<a class="anchor" aria-label="anchor" href="#arg-sizevec"></a></dt>
<dd><p>A vector of total read-counts for the individuals.</p></dd>


<dt id="arg-geno">geno<a class="anchor" aria-label="anchor" href="#arg-geno"></a></dt>
<dd><p>A vector of genotypes for the individuals. I.e. the number of reference alleles
each individual has.</p></dd>


<dt id="arg-ploidy">ploidy<a class="anchor" aria-label="anchor" href="#arg-ploidy"></a></dt>
<dd><p>The ploidy of the species.</p></dd>


<dt id="arg-seq">seq<a class="anchor" aria-label="anchor" href="#arg-seq"></a></dt>
<dd><p>The sequencing error rate.</p></dd>


<dt id="arg-bias">bias<a class="anchor" aria-label="anchor" href="#arg-bias"></a></dt>
<dd><p>The bias parameter. Pr(a read after selected) / Pr(A read after selected).</p></dd>


<dt id="arg-od">od<a class="anchor" aria-label="anchor" href="#arg-od"></a></dt>
<dd><p>The overdispersion parameter. See the
Details of the <code>rho</code> variable in <code><a href="betabinom.html">betabinom</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A vector the same length as <code>sizevec</code>. The ith element
    is the number of reference counts for individual i.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>

<ul><li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">doi:10.1534/genetics.118.301468</a>
.</p></li>
<li><p>Gerard, David, and Luís Felipe Ventorim Ferrão. "Priors for genotyping polyploids." Bioinformatics 36, no. 6 (2020): 1795-1800. <a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link">doi:10.1093/bioinformatics/btz852</a>
.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="rgeno.html">rgeno</a></code> for a way to generate genotypes of individuals. <code><a href="betabinom.html">rbetabinom</a></code>
    for how we generate the read-counts.</p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>David Gerard</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">n</span>       <span class="op">&lt;-</span> <span class="fl">100</span></span></span>
<span class="r-in"><span><span class="va">ploidy</span>  <span class="op">&lt;-</span> <span class="fl">6</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Generate the genotypes of individuals from an F1 population,</span></span></span>
<span class="r-in"><span><span class="co">## where the first parent has 1 copy of the reference allele</span></span></span>
<span class="r-in"><span><span class="co">## and the second parent has two copies of the reference</span></span></span>
<span class="r-in"><span><span class="co">## allele.</span></span></span>
<span class="r-in"><span><span class="va">genovec</span> <span class="op">&lt;-</span> <span class="fu"><a href="rgeno.html">rgeno</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, model <span class="op">=</span> <span class="st">"f1"</span>,</span></span>
<span class="r-in"><span>                 p1geno <span class="op">=</span> <span class="fl">1</span>, p2geno <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get the total number of read-counts for each individual.</span></span></span>
<span class="r-in"><span><span class="co">## Ideally, you would take this from real data as the total</span></span></span>
<span class="r-in"><span><span class="co">## read-counts are definitely not Poisson.</span></span></span>
<span class="r-in"><span><span class="va">sizevec</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, lambda <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Generate the counts of reads with the reference allele</span></span></span>
<span class="r-in"><span><span class="co">## when there is a strong bias for the reference allele</span></span></span>
<span class="r-in"><span><span class="co">## and there is no overdispersion.</span></span></span>
<span class="r-in"><span><span class="va">refvec</span>  <span class="op">&lt;-</span> <span class="fu">rflexdog</span><span class="op">(</span>sizevec <span class="op">=</span> <span class="va">sizevec</span>, geno <span class="op">=</span> <span class="va">genovec</span>,</span></span>
<span class="r-in"><span>                    ploidy <span class="op">=</span> <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.001</span>,</span></span>
<span class="r-in"><span>                    bias <span class="op">=</span> <span class="fl">0.5</span>, od <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Plot the simulated data using plot_geno.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot_geno.html">plot_geno</a></span><span class="op">(</span>refvec <span class="op">=</span> <span class="va">refvec</span>, sizevec <span class="op">=</span> <span class="va">sizevec</span>,</span></span>
<span class="r-in"><span>          ploidy <span class="op">=</span> <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.001</span>, bias <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="rflexdog-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

