<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Simulate GBS data from the flexdog likelihood. — rflexdog • updog</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate GBS data from the flexdog likelihood. — rflexdog"><meta property="og:description" content="This will take a vector of genotypes and a vector of total read-counts, then generate a vector of reference
counts. To get the genotypes, you could use rgeno. The likelihood used to generate read-counts
is described in detail in Gerard et. al. (2018)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a>
    </li>
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/updog/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Simulate GBS data from the <code><a href="flexdog.html">flexdog</a></code> likelihood.</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/R/rflexdog.R" class="external-link"><code>R/rflexdog.R</code></a></small>
    <div class="hidden name"><code>rflexdog.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This will take a vector of genotypes and a vector of total read-counts, then generate a vector of reference
counts. To get the genotypes, you could use <code><a href="rgeno.html">rgeno</a></code>. The likelihood used to generate read-counts
is described in detail in Gerard et. al. (2018).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">rflexdog</span><span class="op">(</span><span class="va">sizevec</span>, <span class="va">geno</span>, <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.005</span>, bias <span class="op">=</span> <span class="fl">1</span>, od <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>sizevec</dt>
<dd><p>A vector of total read-counts for the individuals.</p></dd>
<dt>geno</dt>
<dd><p>A vector of genotypes for the individuals. I.e. the number of reference alleles
each individual has.</p></dd>
<dt>ploidy</dt>
<dd><p>The ploidy of the species.</p></dd>
<dt>seq</dt>
<dd><p>The sequencing error rate.</p></dd>
<dt>bias</dt>
<dd><p>The bias parameter. Pr(a read after selected) / Pr(A read after selected).</p></dd>
<dt>od</dt>
<dd><p>The overdispersion parameter. See the
Details of the <code>rho</code> variable in <code><a href="betabinom.html">betabinom</a></code>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A vector the same length as <code>sizevec</code>. The ith element
    is the number of reference counts for individual i.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    
<ul><li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. doi: <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">10.1534/genetics.118.301468</a>
.</p></li>
<li><p>Gerard, David, and Luís Felipe Ventorim Ferrão. "Priors for genotyping polyploids." Bioinformatics 36, no. 6 (2020): 1795-1800. doi: <a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link">10.1093/bioinformatics/btz852</a>
.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="rgeno.html">rgeno</a></code> for a way to generate genotypes of individuals. <code><a href="betabinom.html">rbetabinom</a></code>
    for how we generate the read-counts.</p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span class="r-in"><span class="va">n</span>       <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span class="r-in"><span class="va">ploidy</span>  <span class="op">&lt;-</span> <span class="fl">6</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Generate the genotypes of individuals from an F1 population,</span></span>
<span class="r-in"><span class="co">## where the first parent has 1 copy of the reference allele</span></span>
<span class="r-in"><span class="co">## and the second parent has two copies of the reference</span></span>
<span class="r-in"><span class="co">## allele.</span></span>
<span class="r-in"><span class="va">genovec</span> <span class="op">&lt;-</span> <span class="fu"><a href="rgeno.html">rgeno</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, model <span class="op">=</span> <span class="st">"f1"</span>,</span>
<span class="r-in">                 p1geno <span class="op">=</span> <span class="fl">1</span>, p2geno <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Get the total number of read-counts for each individual.</span></span>
<span class="r-in"><span class="co">## Ideally, you would take this from real data as the total</span></span>
<span class="r-in"><span class="co">## read-counts are definitely not Poisson.</span></span>
<span class="r-in"><span class="va">sizevec</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, lambda <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Generate the counts of reads with the reference allele</span></span>
<span class="r-in"><span class="co">## when there is a strong bias for the reference allele</span></span>
<span class="r-in"><span class="co">## and there is no overdispersion.</span></span>
<span class="r-in"><span class="va">refvec</span>  <span class="op">&lt;-</span> <span class="fu">rflexdog</span><span class="op">(</span>sizevec <span class="op">=</span> <span class="va">sizevec</span>, geno <span class="op">=</span> <span class="va">genovec</span>,</span>
<span class="r-in">                    ploidy <span class="op">=</span> <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span class="r-in">                    bias <span class="op">=</span> <span class="fl">0.5</span>, od <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Plot the simulated data using plot_geno.</span></span>
<span class="r-in"><span class="fu"><a href="plot_geno.html">plot_geno</a></span><span class="op">(</span>refvec <span class="op">=</span> <span class="va">refvec</span>, sizevec <span class="op">=</span> <span class="va">sizevec</span>,</span>
<span class="r-in">          ploidy <span class="op">=</span> <span class="va">ploidy</span>, seq <span class="op">=</span> <span class="fl">0.001</span>, bias <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span class="r-plt img"><img src="rflexdog-1.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

