<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Simulate individual genotypes from one of the supported flexdog models. — rgeno • updog</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate individual genotypes from one of the supported flexdog models. — rgeno"><meta name="description" content="This will simulate genotypes of a sample of individuals drawn from one of the populations supported by
flexdog. See the details of flexdog for the models allowed. These genotype
distributions are described in detail in Gerard and Ferrão (2020)."><meta property="og:description" content="This will simulate genotypes of a sample of individuals drawn from one of the populations supported by
flexdog. See the details of flexdog for the models allowed. These genotype
distributions are described in detail in Gerard and Ferrão (2020)."><meta property="og:image" content="https://dcgerard.github.io/updog/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">updog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a></li>
    <li><a class="dropdown-item" href="../articles/oracle_calculations.html">Oracle Calculations</a></li>
    <li><a class="dropdown-item" href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a></li>
    <li><a class="dropdown-item" href="../articles/smells_like_updog.html">Example Use of Updog</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/dcgerard/updog/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Simulate individual genotypes from one of the supported <code><a href="flexdog.html">flexdog</a></code> models.</h1>
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/R/rflexdog.R" class="external-link"><code>R/rflexdog.R</code></a></small>
      <div class="d-none name"><code>rgeno.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This will simulate genotypes of a sample of individuals drawn from one of the populations supported by
<code><a href="flexdog.html">flexdog</a></code>. See the details of <code><a href="flexdog.html">flexdog</a></code> for the models allowed. These genotype
distributions are described in detail in Gerard and Ferrão (2020).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rgeno</span><span class="op">(</span></span>
<span>  <span class="va">n</span>,</span>
<span>  <span class="va">ploidy</span>,</span>
<span>  model <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hw"</span>, <span class="st">"bb"</span>, <span class="st">"norm"</span>, <span class="st">"f1"</span>, <span class="st">"s1"</span>, <span class="st">"flex"</span>, <span class="st">"uniform"</span><span class="op">)</span>,</span>
<span>  allele_freq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  od <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p1geno <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p2geno <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pivec <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mu <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sigma <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-n">n<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>The number of observations.</p></dd>


<dt id="arg-ploidy">ploidy<a class="anchor" aria-label="anchor" href="#arg-ploidy"></a></dt>
<dd><p>The ploidy of the species.</p></dd>


<dt id="arg-model">model<a class="anchor" aria-label="anchor" href="#arg-model"></a></dt>
<dd><p>What form should the prior take? See Details in <code><a href="flexdog.html">flexdog</a></code>.</p></dd>


<dt id="arg-allele-freq">allele_freq<a class="anchor" aria-label="anchor" href="#arg-allele-freq"></a></dt>
<dd><p>If <code>model = "hw"</code>, then this is the allele frequency of the population.
For any other model, this should be <code>NULL</code>.</p></dd>


<dt id="arg-od">od<a class="anchor" aria-label="anchor" href="#arg-od"></a></dt>
<dd><p>If <code>model = "bb"</code>, then this is the overdispersion parameter of the beta-binomial
distribution. See <code><a href="betabinom.html">betabinom</a></code> for details. For any other model, this should be
<code>NULL</code>.</p></dd>


<dt id="arg-p-geno">p1geno<a class="anchor" aria-label="anchor" href="#arg-p-geno"></a></dt>
<dd><p>Either the first parent's genotype if <code>model = "f1"</code>,
or the only parent's
genotype if <code>model = "s1"</code>.
For any other model, this should be <code>NULL</code>.</p></dd>


<dt id="arg-p-geno">p2geno<a class="anchor" aria-label="anchor" href="#arg-p-geno"></a></dt>
<dd><p>The second parent's genotype if <code>model = "f1"</code>.
For any other model, this should be <code>NULL</code>.</p></dd>


<dt id="arg-pivec">pivec<a class="anchor" aria-label="anchor" href="#arg-pivec"></a></dt>
<dd><p>A vector of probabilities. If <code>model = "ash"</code>, then this represents
the mixing proportions of the discrete uniforms. If
<code>model = "flex"</code>, then element <code>i</code> is the probability of genotype <code>i - 1</code>.
For any other model, this should be <code>NULL</code>.</p></dd>


<dt id="arg-mu">mu<a class="anchor" aria-label="anchor" href="#arg-mu"></a></dt>
<dd><p>If <code>model = "norm"</code>, this is the mean of the normal.
For any other model, this should be <code>NULL</code>.</p></dd>


<dt id="arg-sigma">sigma<a class="anchor" aria-label="anchor" href="#arg-sigma"></a></dt>
<dd><p>If <code>model = "norm"</code>, this is the standard deviation of the normal.
For any other model, this should be <code>NULL</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A vector of length <code>n</code> with the genotypes of the sampled individuals.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>List of non-<code>NULL</code> arguments:</p><dl><dt><code>model = "flex"</code>:</dt>
<dd><p><code>pivec</code></p></dd>

  <dt><code>model = "hw"</code>:</dt>
<dd><p><code>allele_freq</code></p></dd>

  <dt><code>model = "f1"</code>:</dt>
<dd><p><code>p1geno</code> and <code>p2geno</code></p></dd>

  <dt><code>model = "s1"</code>:</dt>
<dd><p><code>p1geno</code></p></dd>

  <dt><code>model = "uniform"</code>:</dt>
<dd><p>no non-<code>NULL</code> arguments</p></dd>

  <dt><code>model = "bb"</code>:</dt>
<dd><p><code>allele_freq</code> and <code>od</code></p></dd>

  <dt><code>model == "norm"</code>:</dt>
<dd><p><code>mu</code> and <code>sigma</code></p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>

<ul><li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">doi:10.1534/genetics.118.301468</a>
.</p></li>
<li><p>Gerard, David, and Luís Felipe Ventorim Ferrão. "Priors for genotyping polyploids." Bioinformatics 36, no. 6 (2020): 1795-1800. <a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link">doi:10.1093/bioinformatics/btz852</a>
.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>David Gerard</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## F1 Population where parent 1 has 1 copy of the referenc allele</span></span></span>
<span class="r-in"><span><span class="co">## and parent 2 has 4 copies of the reference allele.</span></span></span>
<span class="r-in"><span><span class="va">ploidy</span> <span class="op">&lt;-</span> <span class="fl">6</span></span></span>
<span class="r-in"><span><span class="fu">rgeno</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, model <span class="op">=</span> <span class="st">"f1"</span>, p1geno <span class="op">=</span> <span class="fl">1</span>, p2geno <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 3 4 1 3 3 3 2 1 4 3</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## A population in Hardy-Weinberge equilibrium with an</span></span></span>
<span class="r-in"><span><span class="co">## allele frequency of 0.75</span></span></span>
<span class="r-in"><span><span class="fu">rgeno</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, model <span class="op">=</span> <span class="st">"hw"</span>, allele_freq <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 3 5 5 4 4 4 5 4 6 5</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

