<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Simulate individual genotypes from one of the supported flexdog models. — rgeno • updog</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate individual genotypes from one of the supported flexdog models. — rgeno"><meta property="og:description" content="This will simulate genotypes of a sample of individuals drawn from one of the populations supported by
flexdog. See the details of flexdog for the models allowed. These genotype
distributions are described in detail in Gerard and Ferrão (2020)."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a>
    </li>
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/updog/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Simulate individual genotypes from one of the supported <code><a href="flexdog.html">flexdog</a></code> models.</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/R/rflexdog.R" class="external-link"><code>R/rflexdog.R</code></a></small>
    <div class="hidden name"><code>rgeno.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This will simulate genotypes of a sample of individuals drawn from one of the populations supported by
<code><a href="flexdog.html">flexdog</a></code>. See the details of <code><a href="flexdog.html">flexdog</a></code> for the models allowed. These genotype
distributions are described in detail in Gerard and Ferrão (2020).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rgeno</span><span class="op">(</span></span>
<span>  <span class="va">n</span>,</span>
<span>  <span class="va">ploidy</span>,</span>
<span>  model <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hw"</span>, <span class="st">"bb"</span>, <span class="st">"norm"</span>, <span class="st">"f1"</span>, <span class="st">"s1"</span>, <span class="st">"flex"</span>, <span class="st">"uniform"</span><span class="op">)</span>,</span>
<span>  allele_freq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  od <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p1geno <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p2geno <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pivec <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mu <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sigma <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>n</dt>
<dd><p>The number of observations.</p></dd>


<dt>ploidy</dt>
<dd><p>The ploidy of the species.</p></dd>


<dt>model</dt>
<dd><p>What form should the prior take? See Details in <code><a href="flexdog.html">flexdog</a></code>.</p></dd>


<dt>allele_freq</dt>
<dd><p>If <code>model = "hw"</code>, then this is the allele frequency of the population.
For any other model, this should be <code>NULL</code>.</p></dd>


<dt>od</dt>
<dd><p>If <code>model = "bb"</code>, then this is the overdispersion parameter of the beta-binomial
distribution. See <code><a href="betabinom.html">betabinom</a></code> for details. For any other model, this should be
<code>NULL</code>.</p></dd>


<dt>p1geno</dt>
<dd><p>Either the first parent's genotype if <code>model = "f1"</code>,
or the only parent's
genotype if <code>model = "s1"</code>.
For any other model, this should be <code>NULL</code>.</p></dd>


<dt>p2geno</dt>
<dd><p>The second parent's genotype if <code>model = "f1"</code>.
For any other model, this should be <code>NULL</code>.</p></dd>


<dt>pivec</dt>
<dd><p>A vector of probabilities. If <code>model = "ash"</code>, then this represents
the mixing proportions of the discrete uniforms. If
<code>model = "flex"</code>, then element <code>i</code> is the probability of genotype <code>i - 1</code>.
For any other model, this should be <code>NULL</code>.</p></dd>


<dt>mu</dt>
<dd><p>If <code>model = "norm"</code>, this is the mean of the normal.
For any other model, this should be <code>NULL</code>.</p></dd>


<dt>sigma</dt>
<dd><p>If <code>model = "norm"</code>, this is the standard deviation of the normal.
For any other model, this should be <code>NULL</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A vector of length <code>n</code> with the genotypes of the sampled individuals.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>List of non-<code>NULL</code> arguments:</p><dl><dt><code>model = "flex"</code>:</dt>
<dd><p><code>pivec</code></p></dd>

  <dt><code>model = "hw"</code>:</dt>
<dd><p><code>allele_freq</code></p></dd>

  <dt><code>model = "f1"</code>:</dt>
<dd><p><code>p1geno</code> and <code>p2geno</code></p></dd>

  <dt><code>model = "s1"</code>:</dt>
<dd><p><code>p1geno</code></p></dd>

  <dt><code>model = "uniform"</code>:</dt>
<dd><p>no non-<code>NULL</code> arguments</p></dd>

  <dt><code>model = "bb"</code>:</dt>
<dd><p><code>allele_freq</code> and <code>od</code></p></dd>

  <dt><code>model == "norm"</code>:</dt>
<dd><p><code>mu</code> and <code>sigma</code></p></dd>


</dl></div>
    <div id="references">
    <h2>References</h2>
    
<ul><li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">doi:10.1534/genetics.118.301468</a>
.</p></li>
<li><p>Gerard, David, and Luís Felipe Ventorim Ferrão. "Priors for genotyping polyploids." Bioinformatics 36, no. 6 (2020): 1795-1800. <a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link">doi:10.1093/bioinformatics/btz852</a>
.</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## F1 Population where parent 1 has 1 copy of the referenc allele</span></span></span>
<span class="r-in"><span><span class="co">## and parent 2 has 4 copies of the reference allele.</span></span></span>
<span class="r-in"><span><span class="va">ploidy</span> <span class="op">&lt;-</span> <span class="fl">6</span></span></span>
<span class="r-in"><span><span class="fu">rgeno</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, model <span class="op">=</span> <span class="st">"f1"</span>, p1geno <span class="op">=</span> <span class="fl">1</span>, p2geno <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 3 3 3 2 2 4 1 2 2 2</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## A population in Hardy-Weinberge equilibrium with an</span></span></span>
<span class="r-in"><span><span class="co">## allele frequency of 0.75</span></span></span>
<span class="r-in"><span><span class="fu">rgeno</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, ploidy <span class="op">=</span> <span class="va">ploidy</span>, model <span class="op">=</span> <span class="st">"hw"</span>, allele_freq <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 5 3 3 4 3 5 5 4 4 4</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

