<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>updog Flexible Genotyping for Polyploids — updog-package • updog</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="updog Flexible Genotyping for Polyploids — updog-package" />
<meta property="og:description" content="Implements empirical Bayes approaches to genotype
polyploids from next generation sequencing data while
accounting for allele bias, overdispersion, and sequencing
error. The main functions are flexdog()
and multidog(), which allow the specification
of many different genotype distributions. Also provided
are functions to simulate genotypes, rgeno(),
and read-counts, rflexdog(), as well as
functions to calculate oracle genotyping error rates,
oracle_mis(), and correlation with the true
genotypes, oracle_cor(). These latter two
functions are useful for read depth calculations. Run
browseVignettes(package = &quot;updog&quot;) in R for example usage. See
Gerard et al. (2018)
&amp;lt;doi: 10.1534/genetics.118.301468
&amp;gt;
and Gerard and Ferrao (2020)
&amp;lt;doi: 10.1093/bioinformatics/btz852
&amp;gt;
for details on the implemented methods." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">updog</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a>
    </li>
    <li>
      <a href="../articles/oracle_calculations.html">Oracle Calculations</a>
    </li>
    <li>
      <a href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    </li>
    <li>
      <a href="../articles/smells_like_updog.html">Example Use of Updog</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dcgerard/updog/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1><code>updog</code> Flexible Genotyping for Polyploids</h1>
    <small class="dont-index">Source: <a href='https://github.com/dcgerard/updog/blob/master/R/updog.R'><code>R/updog.R</code></a></small>
    <div class="hidden name"><code>updog-package.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Implements empirical Bayes approaches to genotype
polyploids from next generation sequencing data while
accounting for allele bias, overdispersion, and sequencing
error. The main functions are <code><a href='flexdog.html'>flexdog</a>()</code>
and <code><a href='multidog.html'>multidog</a>()</code>, which allow the specification
of many different genotype distributions. Also provided
are functions to simulate genotypes, <code><a href='rgeno.html'>rgeno</a>()</code>,
and read-counts, <code><a href='rflexdog.html'>rflexdog</a>()</code>, as well as
functions to calculate oracle genotyping error rates,
<code><a href='oracle_mis.html'>oracle_mis</a>()</code>, and correlation with the true
genotypes, <code><a href='oracle_cor.html'>oracle_cor</a>()</code>. These latter two
functions are useful for read depth calculations. Run
<code><a href='https://rdrr.io/r/utils/browseVignettes.html'>browseVignettes(package = "updog")</a></code> in R for example usage. See
Gerard et al. (2018)
&lt;doi: <a href='https://doi.org/10.1534/genetics.118.301468'>10.1534/genetics.118.301468</a>
&gt;
and Gerard and Ferrao (2020)
&lt;doi: <a href='https://doi.org/10.1093/bioinformatics/btz852'>10.1093/bioinformatics/btz852</a>
&gt;
for details on the implemented methods.</p>
    </div>



    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The package is named <code>updog</code> for "Using
Parental Data for Offspring Genotyping" because
we originally developed the
method for full-sib populations, but it works
now for more general populations.</p>
<p>Our best competitor is probably the <code>fitPoly</code> package,
which you can check out at
<a href='https://cran.r-project.org/package=fitPoly'>https://cran.r-project.org/package=fitPoly</a>. Though, we think
that <code>updog</code> returns better calibrated measures of uncertainty
when you have next-generation sequencing data.</p>
<p>If you find a bug or want an enhancement, please submit an
issue at <a href='https://github.com/dcgerard/updog/issues'>https://github.com/dcgerard/updog/issues</a>.</p>
    <h2 class="hasAnchor" id="updog-functions"><a class="anchor" href="#updog-functions"></a><code>updog</code> Functions</h2>

    

<dl>
  <dt><code><a href='flexdog.html'>flexdog</a>()</code></dt><dd><p>The main function that
      fits an empirical Bayes approach to genotype polyploids
      from next generation sequencing data.</p></dd>
  <dt><code><a href='multidog.html'>multidog</a>()</code></dt><dd><p>A convenience function for running
      <code><a href='flexdog.html'>flexdog</a>()</code> over many SNPs. This function provides
      support for parallel computing.</p></dd>
  <dt><code><a href='format_multidog.html'>format_multidog</a>()</code></dt><dd><p>Return arrayicized elements from the output of <code><a href='multidog.html'>multidog</a>()</code>.</p></dd>
  <dt><code><a href='filter_snp.html'>filter_snp</a>()</code></dt><dd><p>Filter SNPs based on the output of <code><a href='multidog.html'>multidog</a>()</code></p></dd>
  <dt><code><a href='rgeno.html'>rgeno</a>()</code></dt><dd><p>simulate the genotypes of a sample
      from one of the models allowed in <code><a href='flexdog.html'>flexdog</a>()</code>.</p></dd>
  <dt><code><a href='rflexdog.html'>rflexdog</a>()</code></dt><dd><p>Simulate read-counts from the
      <code><a href='flexdog.html'>flexdog</a>()</code> model.</p></dd>
  <dt><code><a href='plot.flexdog.html'>plot.flexdog</a>()</code></dt><dd><p>Plotting the output of
      <code><a href='flexdog.html'>flexdog</a>()</code>.</p></dd>
  <dt><code><a href='plot.multidog.html'>plot.multidog</a>()</code></dt><dd><p>Plotting the output of
      <code><a href='multidog.html'>multidog</a>()</code>.</p></dd>
  <dt><code><a href='oracle_joint.html'>oracle_joint</a>()</code></dt><dd><p>The joint distribution of the true genotype
      and an oracle estimator.</p></dd>
  <dt><code><a href='oracle_plot.html'>oracle_plot</a>()</code></dt><dd><p>Visualize the output of <code><a href='oracle_joint.html'>oracle_joint</a>()</code>.</p></dd>
  <dt><code><a href='oracle_mis.html'>oracle_mis</a>()</code></dt><dd><p>The oracle misclassification error rate (Bayes rate).</p></dd>
  <dt><code><a href='oracle_cor.html'>oracle_cor</a>()</code></dt><dd><p>Correlation between the true genotype and the oracle estimated genotype.</p></dd>

</dl>

    <h2 class="hasAnchor" id="updog-datasets"><a class="anchor" href="#updog-datasets"></a><code>updog</code> Datasets</h2>

    

<dl>
  <dt><code><a href='snpdat.html'>snpdat</a></code></dt><dd><p>A small example dataset for using
      <code><a href='flexdog.html'>flexdog</a></code>.</p></dd>
  <dt><code><a href='uitdewilligen.html'>uitdewilligen</a></code></dt><dd><p>A small example dataset</p></dd>

</dl>

    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    
<ul>
<li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. doi: <a href='https://doi.org/10.1534/genetics.118.301468'>10.1534/genetics.118.301468</a>
.</p></li>
<li><p>Gerard, David, and Luís Felipe Ventorim Ferrão. "Priors for genotyping polyploids." Bioinformatics 36, no. 6 (2020): 1795-1800. doi: <a href='https://doi.org/10.1093/bioinformatics/btz852'>10.1093/bioinformatics/btz852</a>
.</p></li>
</ul>

    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>David Gerard</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


