<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='Implements empirical Bayes approaches to genotype
polyploids from next generation sequencing data while
accounting for allele bias, overdispersion, and sequencing
error. The main functions are flexdog()
and multidog(), which allow the specification
of many different genotype distributions. Also provided
are functions to simulate genotypes, rgeno(),
and read-counts, rflexdog(), as well as
functions to calculate oracle genotyping error rates,
oracle_mis(), and correlation with the true
genotypes, oracle_cor(). These latter two
functions are useful for read depth calculations. Run
browseVignettes(package = "updog") in R for example usage. See
Gerard et al. (2018)
&amp;lt;doi:10.1534/genetics.118.301468
&amp;gt;
and Gerard and Ferrao (2020)
&amp;lt;doi:10.1093/bioinformatics/btz852
&amp;gt;
for details on the implemented methods.'><title>updog Flexible Genotyping for Polyploids — updog-package • updog</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="updog Flexible Genotyping for Polyploids — updog-package"><meta property="og:description" content='Implements empirical Bayes approaches to genotype
polyploids from next generation sequencing data while
accounting for allele bias, overdispersion, and sequencing
error. The main functions are flexdog()
and multidog(), which allow the specification
of many different genotype distributions. Also provided
are functions to simulate genotypes, rgeno(),
and read-counts, rflexdog(), as well as
functions to calculate oracle genotyping error rates,
oracle_mis(), and correlation with the true
genotypes, oracle_cor(). These latter two
functions are useful for read depth calculations. Run
browseVignettes(package = "updog") in R for example usage. See
Gerard et al. (2018)
&amp;lt;doi:10.1534/genetics.118.301468
&amp;gt;
and Gerard and Ferrao (2020)
&amp;lt;doi:10.1093/bioinformatics/btz852
&amp;gt;
for details on the implemented methods.'><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">updog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/multidog.html">Genotyping Many SNPs with multidog()</a>
    <a class="dropdown-item" href="../articles/oracle_calculations.html">Oracle Calculations</a>
    <a class="dropdown-item" href="../articles/simulate_ngs.html">Simulate Next-Generation Sequencing Data</a>
    <a class="dropdown-item" href="../articles/smells_like_updog.html">Example Use of Updog</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/dcgerard/updog/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1><code>updog</code> Flexible Genotyping for Polyploids</h1>
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/updog/blob/HEAD/R/updog.R" class="external-link"><code>R/updog.R</code></a></small>
      <div class="d-none name"><code>updog-package.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Implements empirical Bayes approaches to genotype
polyploids from next generation sequencing data while
accounting for allele bias, overdispersion, and sequencing
error. The main functions are <code><a href="flexdog.html">flexdog</a>()</code>
and <code><a href="multidog.html">multidog</a>()</code>, which allow the specification
of many different genotype distributions. Also provided
are functions to simulate genotypes, <code><a href="rgeno.html">rgeno</a>()</code>,
and read-counts, <code><a href="rflexdog.html">rflexdog</a>()</code>, as well as
functions to calculate oracle genotyping error rates,
<code><a href="oracle_mis.html">oracle_mis</a>()</code>, and correlation with the true
genotypes, <code><a href="oracle_cor.html">oracle_cor</a>()</code>. These latter two
functions are useful for read depth calculations. Run
<code>browseVignettes(package = "updog")</code> in R for example usage. See
Gerard et al. (2018)
&lt;<a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">doi:10.1534/genetics.118.301468</a>
&gt;
and Gerard and Ferrao (2020)
&lt;<a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link">doi:10.1093/bioinformatics/btz852</a>
&gt;
for details on the implemented methods.</p>
    </div>


    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The package is named <code>updog</code> for "Using
Parental Data for Offspring Genotyping" because
we originally developed the
method for full-sib populations, but it works
now for more general populations.</p>
<p>Our best competitor is probably the <code>fitPoly</code> package,
which you can check out at
<a href="https://cran.r-project.org/package=fitPoly" class="external-link">https://cran.r-project.org/package=fitPoly</a>. Though, we think
that <code>updog</code> returns better calibrated measures of uncertainty
when you have next-generation sequencing data.</p>
<p>If you find a bug or want an enhancement, please submit an
issue at <a href="https://github.com/dcgerard/updog/issues" class="external-link">https://github.com/dcgerard/updog/issues</a>.</p>
    </div>
    <div class="section level2">
    <h2 id="updog-functions"><code>updog</code> Functions<a class="anchor" aria-label="anchor" href="#updog-functions"></a></h2>
    

<dl><dt><code><a href="flexdog.html">flexdog</a>()</code></dt>
<dd><p>The main function that
      fits an empirical Bayes approach to genotype polyploids
      from next generation sequencing data.</p></dd>

  <dt><code><a href="multidog.html">multidog</a>()</code></dt>
<dd><p>A convenience function for running
      <code><a href="flexdog.html">flexdog</a>()</code> over many SNPs. This function provides
      support for parallel computing.</p></dd>

  <dt><code><a href="format_multidog.html">format_multidog</a>()</code></dt>
<dd><p>Return arrayicized elements from the output of <code><a href="multidog.html">multidog</a>()</code>.</p></dd>

  <dt><code><a href="filter_snp.html">filter_snp</a>()</code></dt>
<dd><p>Filter SNPs based on the output of <code><a href="multidog.html">multidog</a>()</code></p></dd>

  <dt><code><a href="rgeno.html">rgeno</a>()</code></dt>
<dd><p>simulate the genotypes of a sample
      from one of the models allowed in <code><a href="flexdog.html">flexdog</a>()</code>.</p></dd>

  <dt><code><a href="rflexdog.html">rflexdog</a>()</code></dt>
<dd><p>Simulate read-counts from the
      <code><a href="flexdog.html">flexdog</a>()</code> model.</p></dd>

  <dt><code><a href="plot.flexdog.html">plot.flexdog</a>()</code></dt>
<dd><p>Plotting the output of
      <code><a href="flexdog.html">flexdog</a>()</code>.</p></dd>

  <dt><code><a href="plot.multidog.html">plot.multidog</a>()</code></dt>
<dd><p>Plotting the output of
      <code><a href="multidog.html">multidog</a>()</code>.</p></dd>

  <dt><code><a href="oracle_joint.html">oracle_joint</a>()</code></dt>
<dd><p>The joint distribution of the true genotype
      and an oracle estimator.</p></dd>

  <dt><code><a href="oracle_plot.html">oracle_plot</a>()</code></dt>
<dd><p>Visualize the output of <code><a href="oracle_joint.html">oracle_joint</a>()</code>.</p></dd>

  <dt><code><a href="oracle_mis.html">oracle_mis</a>()</code></dt>
<dd><p>The oracle misclassification error rate (Bayes rate).</p></dd>

  <dt><code><a href="oracle_cor.html">oracle_cor</a>()</code></dt>
<dd><p>Correlation between the true genotype and the oracle estimated genotype.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="updog-datasets"><code>updog</code> Datasets<a class="anchor" aria-label="anchor" href="#updog-datasets"></a></h2>
    

<dl><dt><code><a href="snpdat.html">snpdat</a></code></dt>
<dd><p>A small example dataset for using
      <code><a href="flexdog.html">flexdog</a></code>.</p></dd>

  <dt><code><a href="uitdewilligen.html">uitdewilligen</a></code></dt>
<dd><p>A small example dataset</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    
<ul><li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping Polyploids from Messy Sequencing Data. <em>Genetics</em>, 210(3), 789-807. <a href="https://doi.org/10.1534/genetics.118.301468" class="external-link">doi:10.1534/genetics.118.301468</a>
.</p></li>
<li><p>Gerard, David, and Luís Felipe Ventorim Ferrão. "Priors for genotyping polyploids." Bioinformatics 36, no. 6 (2020): 1795-1800. <a href="https://doi.org/10.1093/bioinformatics/btz852" class="external-link">doi:10.1093/bioinformatics/btz852</a>
.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>David Gerard</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by David Gerard.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

